{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import coral as cr\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "Example case: swapping in a new gene coding sequence into a yeast plasmid.\n",
    "\n",
    "Forward design steps:\n",
    "1. Acquire gene coding sequence of interest\n",
    "2. Make a mockup of the plasmid you want to build\n",
    "3. Design the Gibson primers necessary to build the plasmid\n",
    "\n",
    "Validation of design ('reverse' process):\n",
    "1. Simulate the PCR on the templates to be used, generates fragments.\n",
    "2. Simulate the Gibson reaction on the fragments, generates plasmid.\n",
    "3. Compare simulated constructed plasmid to mockup.\n",
    "\n",
    "Aquarium submission process:\n",
    "1. Log into rehearsal server\n",
    "2. Locate templates for PCR (thing that already exists on Aquarium)\n",
    "3. Submit primers to Aquarium (save the IDs!)\n",
    "4. Submit fragments to Aquarium using template + fragment IDs\n",
    "5. Check your submissions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Forward Design\n",
    "\n",
    "### 1: Acquire gene coding sequence of interest\n",
    "\n",
    "This code comes from yeastmine.yeastgenome.org's query builder. You can run it by changing just the yeast systematic name (starts with Y, ends with C or W). If you want to build a similar query yourself from scratch, use the Query Builder to set one or more constrains and to show one or more columns, then scroll down and click on the 'Python' link."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S000003551 YJL014W S. cerevisiae CCT3 Chaperonin Containing TCP-1 ATGCAAGCTCCAGTGGTATTTATGAATGCTTCTCAAGAGAGAACTACCGGTCGTCAGGCCCAAATCTCCAATATCACTGCTGCAAAGGCAGTTGCCGATGTCATTCGAACTTGCTTAGGTCCAAAAGCTATGTTAAAGATGTTATTGGATCCTATGGGCGGTCTTGTGTTGACTAACGATGGCCACGCTATTTTAAGGGAAATTGATGTTGCACATCCAGCAGCTAAATCAATGTTGGAGTTATCCCGTACTCAAGACGAAGAGGTTGGTGACGGGACTACAACAGTTATTATTCTAGCTGGTGAGATTTTAGCTCAGTGTGCACCTTATTTGATTGAAAAAAATATTCACCCTGTTATTATTATCCAAGCCTTGAAGAAAGCACTGACTGATGCACTAGAAGTTATCAAACAAGTAAGTAAACCTGTCGATGTGGAAAATGATGCCGCTATGAAAAAATTGATTCAAGCCTCTATCGGTACTAAATATGTCATACATTGGTCAGAGAAAATGTGTGAATTAGCTCTAGACGCTGTTAAGACTGTCCGTAAAGACCTGGGACAAACTGTGGAAGGTGAACCTAATTTTGAAATCGATATCAAAAGATATGTCCGTGTGGAGAAGATTCCTGGGGGGGATGTTCTAGATTCTCGCGTCTTAAAGGGTGTCTTATTGAACAAGGATGTTGTTCATCCAAAGATGTCCCGTCACATAGAGAATCCACGTGTTGTTCTTTTGGATTGTCCACTAGAATATAAGAAGGGTGAATCTCAAACCAACATAGAAATTGAAAAGGAGGAAGATTGGAATAGGATTTTACAAATCGAGGAAGAACAAGTTCAGCTAATGTGTGAACAAATTTTAGCCGTTAGGCCAACACTAGTCATCACTGAAAAAGGTGTTTCAGATCTAGCACAGCATTATTTACTAAAGGGTGGATGTAGTGTTTTAAGAAGAGTGAAGAAATCAGACAATAATAGAATTGCTCGTGTAACTGGTGCAACCATTGTCAATCGTGTAGAGGACTTGAAGGAAAGTGATGTCGGTACAAACTGTGGATTATTCAAAGTAGAAATGATTGGGGACGAATATTTCTCATTTTTGGATAACTGCAAGGAACCAAAAGCTTGTACCATAATGTTACGTGGTGGTTCCAAAGATATATTAAACGAAATAGATCGTAACCTACAAGATGCTATGGCTGTTGCTCGTAATGTGATGTTGAGTCCTTCCTTGTCCCCAGGTGGTGGTGCCACCGAGATGGCAGTTTCTGTCAAATTAGCAGAAAAGGCAAAGCAATTAGAAGGGATCCAGCAATGGCCATACCAAGCTGTAGCTGATGCTATGGAGTGTATTCCACGTACTTTGATACAAAATGCAGGTGGTGATCCTATTAGACTCCTATCCCAACTAAGAGCCAAACACGCACAGGGAAATTTTACAACAGGTATTGATGGTGACAAGGGGAAAATTGTAGATATGGTTAGTTACGGTATTTGGGAACCTGAAGTCATCAAGCAGCAAAGTGTCAAAACCGCTATTGAAAGTGCATGTTTACTATTAAGGGTTGATGATATTGTTAGTGGTGTAAGAAAACAAGAATAA\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "\n",
    "# This is an automatically generated script to run your query\n",
    "# to use it you will require the intermine python client.\n",
    "# To install the client, run the following command from a terminal:\n",
    "#\n",
    "#     sudo easy_install intermine\n",
    "#\n",
    "# For further documentation you can visit:\n",
    "#     http://intermine.readthedocs.org/en/latest/web-services/\n",
    "\n",
    "# The following two lines will be needed in every python script:\n",
    "from intermine.webservice import Service\n",
    "service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "\n",
    "# Get a new query on the class (table) you will be querying:\n",
    "query = service.new_query(\"Gene\")\n",
    "\n",
    "# The view specifies the output columns\n",
    "query.add_view(\n",
    "    \"primaryIdentifier\", \"secondaryIdentifier\", \"organism.shortName\", \"symbol\",\n",
    "    \"name\", \"sequence.residues\"\n",
    ")\n",
    "\n",
    "# Uncomment and edit the line below (the default) to select a custom sort order:\n",
    "# query.add_sort_order(\"Gene.primaryIdentifier\", \"ASC\")\n",
    "\n",
    "# You can edit the constraint values below\n",
    "query.add_constraint(\"secondaryIdentifier\", \"=\", \"YJL014W\", code = \"A\")\n",
    "\n",
    "# Uncomment and edit the code below to specify your own custom logic:\n",
    "# query.set_logic(\"A\")\n",
    "\n",
    "for row in query.rows():\n",
    "    print row[\"primaryIdentifier\"], row[\"secondaryIdentifier\"], row[\"organism.shortName\"], \\\n",
    "        row[\"symbol\"], row[\"name\"], row[\"sequence.residues\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ATGCAAGCTCCAGTGGTATTTATGAATGCTTCTCAAGAGA ... GGTTGATGATATTGTTAGTGGTGTAAGAAAACAAGAATAA\n",
       "TACGTTCGAGGTCACCATAAATACTTACGAAGAGTTCTCT ... CCAACTACTATAACAATCACCACATTCTTTTGTTCTTATT"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the first value as a dictionary and make it into a\n",
    "# nice coral.DNA instance.\n",
    "my_gene = dict(query.rows().next())\n",
    "gene = cr.DNA(my_gene['Gene.sequence.residues'])\n",
    "gene.name = my_gene['Gene.secondaryIdentifier']\n",
    "gene_feature = cr.Feature(gene.name, 0, len(gene), 'CDS')\n",
    "gene.features.append(gene_feature)\n",
    "gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2: Make a mockup of the plasmid you want to build"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'coral.sequence._dna.DNA'>\n",
      "5858\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div id=sequence_e09671b4-c47f-11e5-9eaa-02591d3bb8ad></div>\n",
       "        <script>\n",
       "        require(['//d3js.org/d3.v3.min.js'], function(lib) {\n",
       "            window.data = {\"features\": [{\"color\": [\"#b7e6d7\"], \"start\": 0, \"stop\": 24, \"name\": \"PP2\", \"type\": \"primer_bind\"}, {\"color\": [\"#c7b0e3\"], \"start\": 24, \"stop\": 679, \"name\": \"Actual GPD Promoter\", \"type\": \"promoter\"}, {\"color\": [\"#84b0dc\"], \"start\": 679, \"stop\": 707, \"name\": \"PS\", \"type\": \"primer_bind\"}, {\"color\": [\"#ff2837\"], \"start\": 707, \"stop\": 1412, \"name\": \"mCherry\", \"type\": \"CDS\"}, {\"color\": [\"#f8d3a9\"], \"start\": 1412, \"stop\": 1434, \"name\": \"TP\", \"type\": \"misc_feature\"}, {\"color\": [\"#f58a5e\"], \"start\": 1434, \"stop\": 1674, \"name\": \"CYC1_terminator\", \"type\": \"terminator\"}, {\"color\": [\"#b1ff67\"], \"start\": 1443, \"stop\": 1462, \"name\": \"pYESTrp_rev_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#f8d3a9\"], \"start\": 1443, \"stop\": 1462, \"name\": \"CYC1_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#9eafd2\"], \"start\": 1674, \"stop\": 1696, \"name\": \"TS\", \"type\": \"primer_bind\"}, {\"color\": [\"#ffef86\"], \"start\": 1696, \"stop\": 2196, \"name\": \"URA3 3' UTR\", \"type\": \"3'UTR\"}, {\"color\": [\"#b7e6d7\"], \"start\": 2196, \"stop\": 2204, \"name\": \"PmeI(1)\", \"type\": \"misc_feature\"}, {\"color\": [\"#d6b295\"], \"start\": 2208, \"stop\": 2225, \"name\": \"M13 Reverse primer\", \"type\": \"promoter\"}, {\"color\": [\"#c6c9d1\"], \"start\": 2221, \"stop\": 2244, \"name\": \"M13_pUC_rev_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#f8d3a9\"], \"start\": 2263, \"stop\": 2264, \"name\": \"mutant?\", \"type\": \"misc_feature\"}, {\"color\": [\"#75c6a9\"], \"start\": 2595, \"stop\": 3215, \"name\": \"pBR322_origin\", \"type\": \"rep_origin\"}, {\"color\": [\"#85dae9\"], \"start\": 2623, \"stop\": 2624, \"name\": \"mut?\", \"type\": \"misc_feature\"}, {\"color\": [\"#75c6a9\"], \"start\": 3369, \"stop\": 4230, \"name\": \"Ampicillin\", \"type\": \"CDS\"}, {\"color\": [\"#f8d3a9\"], \"start\": 4271, \"stop\": 4300, \"name\": \"AmpR_promoter\", \"type\": \"promoter\"}, {\"color\": [\"#c7b0e3\"], \"start\": 4458, \"stop\": 4481, \"name\": \"pGEX_3_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#85dae9\"], \"start\": 4616, \"stop\": 4624, \"name\": \"PmeI\", \"type\": \"misc_feature\"}, {\"color\": [\"#b1ff67\"], \"start\": 4624, \"stop\": 4872, \"name\": \"URA3 promoter (kind of)\", \"type\": \"promoter\"}, {\"color\": [\"#b4abac\"], \"start\": 4872, \"stop\": 5577, \"name\": \"URA3\", \"type\": \"gene\"}, {\"color\": [\"#85dae9\"], \"start\": 5572, \"stop\": 5573, \"name\": \"mutated to A in ura3-1 strains\", \"type\": \"misc_difference\"}, {\"color\": [\"#faac61\"], \"start\": 5676, \"stop\": 5858, \"name\": \"ADH1 Terminator\", \"type\": \"terminator\"}], \"bottom\": \"GAGGTGTGGTCAATAAGAGCGACCTCATGCTATACCTGAGAAAGCAACCTGACCTACAGGAAAGAGTTACTCAAGAATAAGAATTTTCGTTTTAAAACCTAAGAGTCACTTTAAAATTTGTATACACTTATTTTTTTTATAACTTATTTAATAATAAAAATCATAAATCATAAGAAATTCGCTTAGTTTTGCTGGCCGCATCTTCTCAAATATGCTTCCCAGCCTGCTTTTCTGTAACGTTCACCCTCTACCTTAGCATCCCTTCCCTTTGCAAATAGTCCTCTTCCAACAATAATAATGTCAGATCCTGTAGAGACCACATCATCCACGGTTCTATACTGTTGACCCAATGCGTCTCCCTTGTCATCTAAACCCACACCGGGTGTCATAATCAACCAATCGTAACCTTCATCTCTTCCACCCATGTCTCTTTGAGCAATAAAGCCGATAACAAAATCTTTGTCGCTCTTCGCAATGTCAACAGTACCCTTAGTATATTCTCCAGTAGCTAGGGAGCCCTTGCATGACAATTCTGCTAACATCAAAAGGCCTCTAGGTTCCTTTGTTACTTCTTCCGCCGCCTGCTTCAAACCGCTAACAATACCTGGGCCCACCACACCGTGTGCATTCGTAATGTCTGCCCATTCTGCTATTCTGTATACACCCGCAGAGTACTGCAATTTGACTGTATTACCAATGTCAGCAAATTTTCTGTCTTCGAAGAGTAAAAAATTGTACTTGGCGGATAATGCCTTTAGCGGCTTAACTGTGCCCTCCATGGAAAAATCAGTCAAGATATCCACATGTGTTTTTAGTAAACAAATTTTGGGACCTAATGCTTCAACTAACTCCAGTAATTCCTTGGTGGTACGAACATCCAATGAAGCACACAAGTTTGTTTGCTTTTCGTGCATGATATTAAATAGCTTGGCAGCAACAGGACTAGGATGAGTAGCAGCACGTTCCTTATATGTAGCTTTCGACATGATTTATCTTCGTTTCCTGCAGGTTTTTGTTCTGTGCAGTTGGGTTAAGAATACTGGGCAATTTCATGTTTCTTCAACACCACATATGCGTATATATACCAATCTAAGTCTGTGCTCCTTCCTTCGTTCTTCCTTCTGCTCGGAGATTACCGAATCAAAAAAATTTCAAAGAAACCGGAATCAAAAAAAAGAACAAAAAAAAAAAAGATGAATTGAAAAGCTTTATGGACCCTGAAACCACAGCCACAGTTTAAACTATGGTGCACTCTCAGTACAATCTGCTCTGATGCCGCATAGTTAAGCCAGCCCCGACACCCGCCAACACCCGCTGACGCGCCCTGACGGGCTTGTCTGCTCCCGGCATCCGCTTACAGACAAGCTGTGACCGTCTCCGGGAGCTGCATGTGTCAGAGGTTTTCACCGTCATCACCGAAACGCGCGAGACGAAAGGGCCTCGTGATACGCCTATTTTTATAGGTTAATGTCATGATAATAATGGTTTCTTAGACGTCAGGTGGCACTTTTCGGGGAAATGTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATTCAAATATGTATCCGCTCATGAGACAATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAGTATGAGTATTCAACATTTCCGTGTCGCCCTTATTCCCTTTTTTGCGGCATTTTGCCTTCCTGTTTTTGCTCACCCAGAAACGCTGGTGAAAGTAAAAGATGCTGAAGATCAGTTGGGTGCACGAGTGGGTTACATCGAACTGGATCTCAACAGCGGTAAGATCCTTGAGAGTTTTCGCCCCGAAGAACGTTTTCCAATGATGAGCACTTTTAAAGTTCTGCTATGTGGCGCGGTATTATCCCGTATTGACGCCGGGCAAGAGCAACTCGGTCGCCGCATACACTATTCTCAGAATGACTTGGTTGAGTACTCACCAGTCACAGAAAAGCATCTTACGGATGGCATGACAGTAAGAGAATTATGCAGTGCTGCCATAACCATGAGTGATAACACTGCGGCCAACTTACTTCTGACAACGATCGGAGGACCGAAGGAGCTAACCGCTTTTTTTCACAACATGGGGGATCATGTAACTCGCCTTGATCGTTGGGAACCGGAGCTGAATGAAGCCATACCAAACGACGAGCGTGACACCACGATGCCTGTAGCAATGGCAACAACGTTGCGCAAACTATTAACTGGCGAACTACTTACTCTAGCTTCCCGGCAACAATTAATAGACTGGATGGAGGCGGATAAAGTTGCAGGACCACTTCTGCGCTCGGCCCTTCCGGCTGGCTGGTTTATTGCTGATAAATCTGGAGCCGGTGAGCGTGGGTCTCGCGGTATCATTGCAGCACTGGGGCCAGATGGTAAGCCCTCCCGTATCGTAGTTATCTACACGACGGGCAGTCAGGCAACTATGGATGAACGAAATAGACAGATCGCTGAGATAGGTGCCTCACTGATTAAGCATTGGTAACTGTCAGACCAAGTTTACTCATATATACTTTAGATTGATTTAAAACTTCATTTTTAATTTAAAAGGATCTAGGTGAAGATCCTTTTTGATAATCTCATGACCAAAATCCCTTAACGTGAGTTTTCGTTCCACTGAGCGTCAGACCCCGTAGAAAAGATCAAAGGATCTTCTTGAGATCCTTTTTTTCTGCGCGTAATCTGCTGCTTGCAAACAAAAAAACCACCGCTACCAGCGGTGGTTTGTTTGCCGGATCAAGAGCTACCAACTCTTTTTCCGAAGGTAACTGGCTTCAGCAGAGCGCAGATACCAAATACTGTCCTTCTAGTGTAGCCGTAGTTAGGCCACCACTTCAAGAACTCTGTAGCACCGCCTACATACCTCGCTCTGCTAATCCTGTTACCAGTGGCTGCTGCCAGTGGCGATAAGTCGTGTCTTACCGGGTTGGACTCAAGACGATAGTTACCGGATAAGGCGCAGCGGTCGGGCTGAACGGGGGGTTCGTGCACACAGCCCAGCTTGGAGCGAACGACCTACACCGAACTGAGATACCTACAGCGTGAGCATTGAGAAAGCGCCACGCTTCCCGAAGGGAGAAAGGCGGACAGGTATCCGGTAAGCGGCAGGGTCGGAACAGGAGAGCGCACGAGGGAGCTTCCAGGGGGGAACGCCTGGTATCTTTATAGTCCTGTCGGGTTTCGCCACCTCTGACTTGAGCGTCGATTTTTGTGATGCTCGTCAGGGGGGCGGAGCCTATGGAAAAACGCCAGCAACGCGGCCTTTTTACGGTTCCTGGCCTTTTGCTGGCCTTTTGCTCACATGTTCTTTCCTGCGTTATCCCCTGATTCTGTGGATAACCGTATTACCGCCTTTGAGTGAGCTGATACCGCTCGCCGCAGCCGAACGACCGAGCGCAGCGAGTCAGTGAGCGAGGAAGCGGAAGAGCGCCCAATACGCAAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTGAGCGCAACGCAATTAATGTGAGTTACCTCACTCATTAGGCACCCCAGGCTTTACACTTTATGCTTCCGGCTCGTATGTTGTGTGGAATTGTGAGCGGATAACAATTTCACACAGGAAACAGCTATGACCATGGTTTAAACCGGCCCAAGCCTTGTCCCAAGGCAGCGTTTTGTTCTTGGAAACGCTGCCCTACACGTTCGCTATGCTTCAAGAACTTTTCTGAGCACTTCATGATGCATGTTTGTTCCTTATTGGTTAGCTTTGATGTTGTGAAGTCATTGACACAGTCTGTGAAACATCTTTCTACCAGATTAGAGTACAAACGCATGAAATCCTTCATTTGCTTTTGTTCCACTACTTTTTGGAACTCTTGTTGTTCTTTGGAGTTCAATGCGTCCATCTTTACAGTCCTGTCTTATTGTTCTTGATTTGTGCCCCGTAAAATACTGTTACTTGGTTCTGGCGAGGTATTGGATAGTTCCTTTTTATAAAGGCCATGAAGCTTTTTCTTTCCAATTTTTTTTTTTTCGTCATTATAGAAATCATTACGACCGAGATTCCCGGGTAATAACTGATATAATTAAATTGAAGCTCTAATTTGTGAGTTTAGTATACATGCATTTACTTATAATACAGTTTTCGGGGCGGGGTTATTACGACATCAAATTAAAGCCTTCGAGCGTCCCAAAACCTTCTCAAGCAAGGTTTTCAGTATAATGTTACATGCGTACACGCGTCTGTACAGAAAAAAAAGAAAAATTTGAAATATAAATAACGTTCTTAATACTAACATAACTATAAAAAAATAAATAGGGACCTAGACTTCAGGTTGTCTAACTCCTTCCTTTTCGGTTAGAGCGGATGTGGGGGGAGGGCGTGAATGTAAGCGTGACATAACTAATTGACTCGAGGTCGACGGTATCACTTGTACAGCTCGTCCATGCCGCCGGTGGAGTGGCGGCCCTCGGCGCGTTCGTACTGTTCCACGATGGTGTAGTCCTCGTTGTGGGAGGTGATGTCCAACTTGATGTTGACGTTGTAGGCGCCGGGCAGCTGCACGGGCTTCTTGGCCTTGTAGGTGGTCTTGACCTCAGCGTCGTAGTGGCCGCCGTCCTTCAGCTTCAGCCTCTGCTTGATCTCGCCCTTCAGGGCGCCGTCCTCGGGGTACATCCGCTCGGAGGAGGCCTCCCAGCCCATGGTCTTCTTCTGCATTACGGGGCCGTCGGAGGGGAAGTTGGTGCCGCGCAGCTTCACCTTGTAGATGAACTCGCCGTCCTGCAGGGAGGAGTCCTGGGTCACGGTCACCACGCCGCCGTCCTCGAAGTTCATCACGCGCTCCCACTTGAAGCCCTCGGGGAAGGACAGCTTCAAGTAGTCGGGGATGTCGGCGGGGTGCTTCACGTAGGCCTTGGAGCCGTACATGAACTGAGGGGACAGGATGTCCCAGGCGAAGGGCAGGGGGCCACCCTTGGTCACCTTCAGCTTGGCGGTCTGGGTGCCCTCGTAGGGGCGGCCCTCGCCCTCGCCCTCGATCTCGAACTCGTGGCCGTTCACGGAGCCCTCCATGTGCACCTTGAAGCGCATGAACTCCTTGATGATGGCCATGTTATCCTCCTCGCCCTTGCTCACTTTGTAGATCCACTAGTTCTAGAATCCGATCCGTCGAAACTAAGTTCTGGTGTTTTAAAACTAAAAAAAAGACTAACTATAAAAGTAGAATTTAAGAAGTTTAAGAAATAGATTTACAGAATTACAATCAATACCTACCGTCTTTATATACTTATTAGTCAAGTAGGGGAATAATTTCAGGGAACTGGTTTCAACCTTTTTTTTCAGCTTTTTCCAAATCAGAGAGAGCAGAAGGTAATAGAAGGTGTAAGAAAATGAGATAGATACATGCGTGGGTCAATTGCCTTGTGTCATCATTTACTCCAGGCAGGTTGCATCACTCCATTGAGGTTGTGCCCGTTTTTTGCCTGTTTGTGCCCCTGTTCTCTGTAGTTGCGCTAAGAGAATGGACCTATGAACTGATGGTTGGTGAAGAAAACAATATTTTGGTGCTGGGATTCTTTTTTTTTCTGGATGCCAGCTTAAAAAGCGGGCTCCATTATATTTAGTGGATGCCAGGAATAAACTGTTCACCCAGACACCTACGATGTTATATATTCTGTGTAACCCGCCCCCTATTTTGGGCATGTACGGGTTACAGCAGAATTAAAAGGCTAATTTTTTGACTAAATAAAGTTAGGAAAATCACTACTATTAATTATTTACGTATTCTTTGAAATGGCGAGTATTGATAATGATAAACTGCCGCTTTTCGAATCTGGTTAAGG\", \"material\": \"DNA\", \"sequence\": \"CCTTAACCAGATTCGAAAAGCGGCAGTTTATCATTATCAATACTCGCCATTTCAAAGAATACGTAAATAATTAATAGTAGTGATTTTCCTAACTTTATTTAGTCAAAAAATTAGCCTTTTAATTCTGCTGTAACCCGTACATGCCCAAAATAGGGGGCGGGTTACACAGAATATATAACATCGTAGGTGTCTGGGTGAACAGTTTATTCCTGGCATCCACTAAATATAATGGAGCCCGCTTTTTAAGCTGGCATCCAGAAAAAAAAAGAATCCCAGCACCAAAATATTGTTTTCTTCACCAACCATCAGTTCATAGGTCCATTCTCTTAGCGCAACTACAGAGAACAGGGGCACAAACAGGCAAAAAACGGGCACAACCTCAATGGAGTGATGCAACCTGCCTGGAGTAAATGATGACACAAGGCAATTGACCCACGCATGTATCTATCTCATTTTCTTACACCTTCTATTACCTTCTGCTCTCTCTGATTTGGAAAAAGCTGAAAAAAAAGGTTGAAACCAGTTCCCTGAAATTATTCCCCTACTTGACTAATAAGTATATAAAGACGGTAGGTATTGATTGTAATTCTGTAAATCTATTTCTTAAACTTCTTAAATTCTACTTTTATAGTTAGTCTTTTTTTTAGTTTTAAAACACCAGAACTTAGTTTCGACGGATCGGATTCTAGAACTAGTGGATCTACAAAGTGAGCAAGGGCGAGGAGGATAACATGGCCATCATCAAGGAGTTCATGCGCTTCAAGGTGCACATGGAGGGCTCCGTGAACGGCCACGAGTTCGAGATCGAGGGCGAGGGCGAGGGCCGCCCCTACGAGGGCACCCAGACCGCCAAGCTGAAGGTGACCAAGGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCTCAGTTCATGTACGGCTCCAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACTTGAAGCTGTCCTTCCCCGAGGGCTTCAAGTGGGAGCGCGTGATGAACTTCGAGGACGGCGGCGTGGTGACCGTGACCCAGGACTCCTCCCTGCAGGACGGCGAGTTCATCTACAAGGTGAAGCTGCGCGGCACCAACTTCCCCTCCGACGGCCCCGTAATGCAGAAGAAGACCATGGGCTGGGAGGCCTCCTCCGAGCGGATGTACCCCGAGGACGGCGCCCTGAAGGGCGAGATCAAGCAGAGGCTGAAGCTGAAGGACGGCGGCCACTACGACGCTGAGGTCAAGACCACCTACAAGGCCAAGAAGCCCGTGCAGCTGCCCGGCGCCTACAACGTCAACATCAAGTTGGACATCACCTCCCACAACGAGGACTACACCATCGTGGAACAGTACGAACGCGCCGAGGGCCGCCACTCCACCGGCGGCATGGACGAGCTGTACAAGTGATACCGTCGACCTCGAGTCAATTAGTTATGTCACGCTTACATTCACGCCCTCCCCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAGACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGATGTCGTAATAACCCCGCCCCGAAAACTGTATTATAAGTAAATGCATGTATACTAAACTCACAAATTAGAGCTTCAATTTAATTATATCAGTTATTACCCGGGAATCTCGGTCGTAATGATTTCTATAATGACGAAAAAAAAAAAATTGGAAAGAAAAAGCTTCATGGCCTTTATAAAAAGGAACTATCCAATACCTCGCCAGAACCAAGTAACAGTATTTTACGGGGCACAAATCAAGAACAATAAGACAGGACTGTAAAGATGGACGCATTGAACTCCAAAGAACAACAAGAGTTCCAAAAAGTAGTGGAACAAAAGCAAATGAAGGATTTCATGCGTTTGTACTCTAATCTGGTAGAAAGATGTTTCACAGACTGTGTCAATGACTTCACAACATCAAAGCTAACCAATAAGGAACAAACATGCATCATGAAGTGCTCAGAAAAGTTCTTGAAGCATAGCGAACGTGTAGGGCAGCGTTTCCAAGAACAAAACGCTGCCTTGGGACAAGGCTTGGGCCGGTTTAAACCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAATTCCACACAACATACGAGCCGGAAGCATAAAGTGTAAAGCCTGGGGTGCCTAATGAGTGAGGTAACTCACATTAATTGCGTTGCGCTCACTGCCCGCTTTCCAGTCGGGAAACCTGTCGTGCCAGCTGCATTAATGAATCGGCCAACGCGCGGGGAGAGGCGGTTTGCGTATTGGGCGCTCTTCCGCTTCCTCGCTCACTGACTCGCTGCGCTCGGTCGTTCGGCTGCGGCGAGCGGTATCAGCTCACTCAAAGGCGGTAATACGGTTATCCACAGAATCAGGGGATAACGCAGGAAAGAACATGTGAGCAAAAGGCCAGCAAAAGGCCAGGAACCGTAAAAAGGCCGCGTTGCTGGCGTTTTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAAAATCGACGCTCAAGTCAGAGGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTCCCCCCTGGAAGCTCCCTCGTGCGCTCTCCTGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGGAAGCGTGGCGCTTTCTCAATGCTCACGCTGTAGGTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTGGGCTGTGTGCACGAACCCCCCGTTCAGCCCGACCGCTGCGCCTTATCCGGTAACTATCGTCTTGAGTCCAACCCGGTAAGACACGACTTATCGCCACTGGCAGCAGCCACTGGTAACAGGATTAGCAGAGCGAGGTATGTAGGCGGTGCTACAGAGTTCTTGAAGTGGTGGCCTAACTACGGCTACACTAGAAGGACAGTATTTGGTATCTGCGCTCTGCTGAAGCCAGTTACCTTCGGAAAAAGAGTTGGTAGCTCTTGATCCGGCAAACAAACCACCGCTGGTAGCGGTGGTTTTTTTGTTTGCAAGCAGCAGATTACGCGCAGAAAAAAAGGATCTCAAGAAGATCCTTTGATCTTTTCTACGGGGTCTGACGCTCAGTGGAACGAAAACTCACGTTAAGGGATTTTGGTCATGAGATTATCAAAAAGGATCTTCACCTAGATCCTTTTAAATTAAAAATGAAGTTTTAAATCAATCTAAAGTATATATGAGTAAACTTGGTCTGACAGTTACCAATGCTTAATCAGTGAGGCACCTATCTCAGCGATCTGTCTATTTCGTTCATCCATAGTTGCCTGACTGCCCGTCGTGTAGATAACTACGATACGGGAGGGCTTACCATCTGGCCCCAGTGCTGCAATGATACCGCGAGACCCACGCTCACCGGCTCCAGATTTATCAGCAATAAACCAGCCAGCCGGAAGGGCCGAGCGCAGAAGTGGTCCTGCAACTTTATCCGCCTCCATCCAGTCTATTAATTGTTGCCGGGAAGCTAGAGTAAGTAGTTCGCCAGTTAATAGTTTGCGCAACGTTGTTGCCATTGCTACAGGCATCGTGGTGTCACGCTCGTCGTTTGGTATGGCTTCATTCAGCTCCGGTTCCCAACGATCAAGGCGAGTTACATGATCCCCCATGTTGTGAAAAAAAGCGGTTAGCTCCTTCGGTCCTCCGATCGTTGTCAGAAGTAAGTTGGCCGCAGTGTTATCACTCATGGTTATGGCAGCACTGCATAATTCTCTTACTGTCATGCCATCCGTAAGATGCTTTTCTGTGACTGGTGAGTACTCAACCAAGTCATTCTGAGAATAGTGTATGCGGCGACCGAGTTGCTCTTGCCCGGCGTCAATACGGGATAATACCGCGCCACATAGCAGAACTTTAAAAGTGCTCATCATTGGAAAACGTTCTTCGGGGCGAAAACTCTCAAGGATCTTACCGCTGTTGAGATCCAGTTCGATGTAACCCACTCGTGCACCCAACTGATCTTCAGCATCTTTTACTTTCACCAGCGTTTCTGGGTGAGCAAAAACAGGAAGGCAAAATGCCGCAAAAAAGGGAATAAGGGCGACACGGAAATGTTGAATACTCATACTCTTCCTTTTTCAATATTATTGAAGCATTTATCAGGGTTATTGTCTCATGAGCGGATACATATTTGAATGTATTTAGAAAAATAAACAAATAGGGGTTCCGCGCACATTTCCCCGAAAAGTGCCACCTGACGTCTAAGAAACCATTATTATCATGACATTAACCTATAAAAATAGGCGTATCACGAGGCCCTTTCGTCTCGCGCGTTTCGGTGATGACGGTGAAAACCTCTGACACATGCAGCTCCCGGAGACGGTCACAGCTTGTCTGTAAGCGGATGCCGGGAGCAGACAAGCCCGTCAGGGCGCGTCAGCGGGTGTTGGCGGGTGTCGGGGCTGGCTTAACTATGCGGCATCAGAGCAGATTGTACTGAGAGTGCACCATAGTTTAAACTGTGGCTGTGGTTTCAGGGTCCATAAAGCTTTTCAATTCATCTTTTTTTTTTTTGTTCTTTTTTTTGATTCCGGTTTCTTTGAAATTTTTTTGATTCGGTAATCTCCGAGCAGAAGGAAGAACGAAGGAAGGAGCACAGACTTAGATTGGTATATATACGCATATGTGGTGTTGAAGAAACATGAAATTGCCCAGTATTCTTAACCCAACTGCACAGAACAAAAACCTGCAGGAAACGAAGATAAATCATGTCGAAAGCTACATATAAGGAACGTGCTGCTACTCATCCTAGTCCTGTTGCTGCCAAGCTATTTAATATCATGCACGAAAAGCAAACAAACTTGTGTGCTTCATTGGATGTTCGTACCACCAAGGAATTACTGGAGTTAGTTGAAGCATTAGGTCCCAAAATTTGTTTACTAAAAACACATGTGGATATCTTGACTGATTTTTCCATGGAGGGCACAGTTAAGCCGCTAAAGGCATTATCCGCCAAGTACAATTTTTTACTCTTCGAAGACAGAAAATTTGCTGACATTGGTAATACAGTCAAATTGCAGTACTCTGCGGGTGTATACAGAATAGCAGAATGGGCAGACATTACGAATGCACACGGTGTGGTGGGCCCAGGTATTGTTAGCGGTTTGAAGCAGGCGGCGGAAGAAGTAACAAAGGAACCTAGAGGCCTTTTGATGTTAGCAGAATTGTCATGCAAGGGCTCCCTAGCTACTGGAGAATATACTAAGGGTACTGTTGACATTGCGAAGAGCGACAAAGATTTTGTTATCGGCTTTATTGCTCAAAGAGACATGGGTGGAAGAGATGAAGGTTACGATTGGTTGATTATGACACCCGGTGTGGGTTTAGATGACAAGGGAGACGCATTGGGTCAACAGTATAGAACCGTGGATGATGTGGTCTCTACAGGATCTGACATTATTATTGTTGGAAGAGGACTATTTGCAAAGGGAAGGGATGCTAAGGTAGAGGGTGAACGTTACAGAAAAGCAGGCTGGGAAGCATATTTGAGAAGATGCGGCCAGCAAAACTAAGCGAATTTCTTATGATTTATGATTTTTATTATTAAATAAGTTATAAAAAAAATAAGTGTATACAAATTTTAAAGTGACTCTTAGGTTTTAAAACGAAAATTCTTATTCTTGAGTAACTCTTTCCTGTAGGTCAGGTTGCTTTCTCAGGTATAGCATGAGGTCGCTCTTATTGACCACACCTC\", \"circular\": true, \"name\": \"pMOD6-pGPD-mCherry\"};// Full JSON description of plasmid - in the future, this would be\n",
       "// downloaded dynamically\n",
       "function d3sequence(sequence_json, div_id) {\n",
       "  // TODO: Have methods for drawing circular vs. linear, keep core logic the same\n",
       "  // Option that might make it into a configurable version\n",
       "  // TODO: control more of these with CSS\n",
       "\n",
       "  var circularConfig = {\n",
       "    'width': 500,\n",
       "    'height': 500,\n",
       "    'padding': 75,\n",
       "    'featureScale': 1.2,\n",
       "    'featureOpacity': 0.5,\n",
       "    'backboneThickness': 0.2,\n",
       "    'backboneColor': '#eee',\n",
       "    'minFeatureSize': 5,\n",
       "    'ticks': 80,\n",
       "    'majorTickInterval': 5,\n",
       "    'tickColorMinor': '#94aab0',\n",
       "    'tickColorMajor': 'black',\n",
       "    'tickLength': 6,\n",
       "    'lineFeatureWidth': 4,\n",
       "    'titleFontSize': '25px'\n",
       "  };\n",
       "\n",
       "  var linearConfig = {\n",
       "    'width': 800,\n",
       "    'height': 150,\n",
       "    'padding': 35,\n",
       "    'featureScale': 1.2,\n",
       "    'featureOpacity': 0.5,\n",
       "    'backboneThickness': 0.2,\n",
       "    'backboneColor': '#eee',\n",
       "    'minFeatureSize': 5,\n",
       "    'ticks': 40,\n",
       "    'majorTickInterval': 5,\n",
       "    'tickColorMinor': '#94aab0',\n",
       "    'tickColorMajor': 'black',\n",
       "    'tickLength': 6,\n",
       "    'lineFeatureWidth': 4,\n",
       "    'titleFontSize': '25px'\n",
       "  };\n",
       "\n",
       "  if (!sequence_json.circular) {\n",
       "    config = linearConfig;\n",
       "  } else if (sequence_json.circular) {\n",
       "    config = circularConfig;\n",
       "  }\n",
       "\n",
       "  ///////////////////////\n",
       "  // Initialize svg space\n",
       "  ///////////////////////\n",
       "  var svg = d3.select('#' + div_id)\n",
       "    .append('svg')\n",
       "      .attr('width', config.width)\n",
       "      .attr('height', config.height)\n",
       "    .append('g');\n",
       "\n",
       "  var center = {'x': config.width / 2, 'y': config.height / 2};\n",
       "\n",
       "  if (sequence_json.circular) {\n",
       "\n",
       "    var outerRadius = Math.min(config.width, config.height) / 2 - config.padding;\n",
       "    var innerRadius = outerRadius - config.backboneThickness * outerRadius;\n",
       "    var centerRadius = d3.mean([outerRadius, innerRadius]);\n",
       "    var featureOffset = (config.featureScale * (outerRadius - innerRadius) - (outerRadius - innerRadius)) / 2;\n",
       "\n",
       "    /////////////////////////////////////////\n",
       "    // Placement rules for circular sequences\n",
       "    /////////////////////////////////////////\n",
       "    var centerTranslation = 'translate(' + center.x + ',' + center.y + ')';\n",
       "    svg.attr('transform', centerTranslation);\n",
       "\n",
       "    var baseToRadians = d3.scale.linear()\n",
       "        .domain([0, sequence_json.sequence.length])\n",
       "        .range([-(Math.PI / 2), 1.5 * Math.PI]);\n",
       "\n",
       "    // Placement around the plasmid is always prior to centering in the svg\n",
       "    function polarToCartesian(radius, angle) {\n",
       "      return {'x': radius * Math.cos(angle), 'y': radius * Math.sin(angle)};\n",
       "    }\n",
       "\n",
       "\n",
       "    ///////////////////////////////////////////\n",
       "    // Make an annulus for the plasmid backbone\n",
       "    ///////////////////////////////////////////\n",
       "\n",
       "    var annulus = d3.svg.arc()\n",
       "        .innerRadius(innerRadius)\n",
       "        .outerRadius(outerRadius)\n",
       "        .startAngle(0)\n",
       "        .endAngle(2 * Math.PI);\n",
       "\n",
       "    backbone = svg.append('path')\n",
       "        .attr('class', 'backbone')\n",
       "        .attr('d', annulus)\n",
       "        .style('fill-rule', 'evenodd')\n",
       "        .style('fill', config.backboneColor)\n",
       "        .style('stroke', d3.rgb(config.backboneColor).darker(0.5))\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    /////////////////////////////////\n",
       "    // Add plasmid name, number of bp\n",
       "    /////////////////////////////////\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.name)\n",
       "        .attr('x', 0)\n",
       "        .attr('y', 0)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', config.titleFontSize);\n",
       "\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.sequence.length + ' bp')\n",
       "        .attr('x', 0)\n",
       "        .attr('y', 40)\n",
       "        .style('text-anchor', 'middle');\n",
       "\n",
       "    /////////////\n",
       "    // Tick Marks\n",
       "    /////////////\n",
       "\n",
       "    // Generate ticks data\n",
       "    var tickData = [];\n",
       "    for (var i=0; i < config.ticks; i++) {\n",
       "      tickData.push(i * Math.floor(sequence_json.sequence.length / config.ticks));\n",
       "    }\n",
       "\n",
       "    // FIXME: why use paths to make Ticks when lines can be used?\n",
       "    var line = d3.svg.line()\n",
       "        .x(function(d) { return d.x; } )\n",
       "        .y(function(d) { return d.y; } );\n",
       "\n",
       "    tickGroup = svg.append('g')\n",
       "        .attr('id', 'tick-group');\n",
       "\n",
       "    ticks = tickGroup.selectAll('.tick')\n",
       "        .data(tickData)\n",
       "      .enter().append('path')\n",
       "        .attr('d', function(d, i) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          inner = polarToCartesian(outerRadius, tickAngle);\n",
       "          outer = polarToCartesian(outerRadius + config.tickLength, tickAngle);\n",
       "          return line([inner, outer]);\n",
       "        })\n",
       "        .attr('stroke', function(d, i) {\n",
       "          if (i === 0) {\n",
       "            tickColor = 'red';\n",
       "          } else if (i % config.majorTickInterval === 0) {\n",
       "            tickColor = config.tickColorMajor;\n",
       "          } else {\n",
       "            tickColor = config.tickColorMinor;\n",
       "          }\n",
       "          return tickColor;\n",
       "        })\n",
       "        .attr('stroke-width', 2)\n",
       "        .attr('fill', 'none');\n",
       "\n",
       "    tickLabels = tickGroup.selectAll('.tick-label')\n",
       "        .data(tickData)\n",
       "      .enter().append('text')\n",
       "        .text(function(d) {\n",
       "            if (d % config.majorTickInterval === 0) {\n",
       "              return d;\n",
       "            }\n",
       "        })\n",
       "        .attr('x', function(d) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          return polarToCartesian(outerRadius + 22, tickAngle).x;\n",
       "        })\n",
       "        .attr('y', function(d) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          return polarToCartesian(outerRadius + 22, tickAngle).y;\n",
       "        })\n",
       "        .attr('dy', '0.4em')\n",
       "        .attr('fill', config.tickColorMinor)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', '12px');\n",
       "\n",
       "    ////////////////////////////////\n",
       "    // Add features and their labels\n",
       "    ////////////////////////////////\n",
       "\n",
       "    // Separate out arc features and line features\n",
       "    var blockData = [];\n",
       "    var lineData = [];\n",
       "    for (var i = 0; i < sequence_json.features.length; i++) {\n",
       "      feature = sequence_json.features[i];\n",
       "      arcAngle = 2 * Math.PI / sequence_json.sequence.length * (feature.stop - feature.start);\n",
       "      arcLength = centerRadius * arcAngle;\n",
       "      if (arcLength > config.minFeatureSize) {\n",
       "        blockData.push(feature);\n",
       "      } else {\n",
       "        lineData.push(feature);\n",
       "      }\n",
       "    }\n",
       "\n",
       "    var featuresGroup = svg.append('g')\n",
       "        .attr('id', 'features');\n",
       "    var blockGroup = featuresGroup.append('g')\n",
       "        .attr('id', 'block-features');\n",
       "\n",
       "    var blockFeatures = blockGroup.selectAll('.feature')\n",
       "        .data(blockData)\n",
       "      .enter()\n",
       "        .append('g')\n",
       "        .attr('class', 'feature')\n",
       "        .append('path')\n",
       "        .attr('d', function(d, i) {\n",
       "          the_arc = d3.svg.arc()\n",
       "            // d3 svg arc origin is at the top\n",
       "            .startAngle(baseToRadians(d.start) + Math.PI / 2)\n",
       "            .endAngle(baseToRadians(d.stop) + Math.PI / 2)\n",
       "            .innerRadius(innerRadius - featureOffset)\n",
       "            .outerRadius(outerRadius + featureOffset)\n",
       "            (d, i);\n",
       "          return the_arc;\n",
       "        })\n",
       "        .attr('id', function(d, i) { return 'block' + i; })\n",
       "        .style('fill-rule', 'evenodd')\n",
       "        .style('fill', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            return d.color;\n",
       "          } else {\n",
       "            return 'rgb(192, 192, 256)';\n",
       "          }\n",
       "        })\n",
       "        .style('fill-opacity', config.featureOpacity)\n",
       "        .style('stroke', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            color = d.color;\n",
       "          } else {\n",
       "            color = 'rgb(192, 192, 256)';\n",
       "          }\n",
       "          return d3.rgb(color).darker(0.5);\n",
       "        })\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "\n",
       "    // Gaussian blur filter for making text shadow\n",
       "    var filter = svg.append('defs')\n",
       "      .append('filter')\n",
       "        .attr('id', 'text-shadow')\n",
       "        .attr('height', '130%')\n",
       "      .append('feGaussianBlur')\n",
       "        .attr('in', 'SourceAlpha')\n",
       "        .attr('stdDeviation', 2)\n",
       "        .attr('result', 'blur');\n",
       "\n",
       "    // Add invisible arcs for aligning text\n",
       "    function describeArc(radius, start, end) {\n",
       "      start_xy = polarToCartesian(radius, baseToRadians(start));\n",
       "      end_xy = polarToCartesian(radius, baseToRadians(end));\n",
       "\n",
       "      var d = [\n",
       "        'M', start_xy.x, start_xy.y,\n",
       "        'A', radius, radius, 0, 0, 1, end_xy.x, end_xy.y\n",
       "      ].join(' ');\n",
       "\n",
       "      return d;\n",
       "    };\n",
       "\n",
       "    var centerArcs = blockGroup.selectAll('.feature, .center-arcs')\n",
       "      .append('path')\n",
       "        .attr('d', function(d) {\n",
       "          return describeArc(0.97 * centerRadius, d.start, d.stop)\n",
       "        })\n",
       "        .attr('class', 'center-arc')\n",
       "        .attr('id', function(d, i) { return div_id + '-center-arc' + i; })\n",
       "        .style('stroke', 'none')\n",
       "        .style('fill', 'none');\n",
       "\n",
       "    var shadows = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-label-shadow')\n",
       "      .append('textPath')\n",
       "        .attr('class', 'shadow')\n",
       "        .attr('startOffset', '50%')\n",
       "        .attr('xlink:href', function(d, i) {\n",
       "          return '#' + div_id + '-center-arc' + i;\n",
       "        })\n",
       "        .attr('id', function(d, i) {\n",
       "          return 'shadow' + i;\n",
       "        })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '11px')\n",
       "        .style('filter', 'url(#text-shadow)')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    var blockText = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-label')\n",
       "      .append('textPath')\n",
       "        .attr('class', 'block-text')\n",
       "        .attr('fill', 'white')\n",
       "        .attr('startOffset', '50%')\n",
       "        .attr('xlink:href', function(d, i) {\n",
       "          return '#' + div_id + '-center-arc' + i;\n",
       "        })\n",
       "        .attr('id', function(d, i) { return 'label' + i; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '11px')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    // If text is too big, don't offset and set anchor to left\n",
       "    // FIXME: this should really be done with more d3-style methods\n",
       "    feature_selection = document.getElementsByClassName('feature');\n",
       "    for (var i = 0; i < feature_selection.length; i++) {\n",
       "      blocktemp = feature_selection[i];\n",
       "      text = blocktemp.getElementsByClassName('block-text')[0];\n",
       "      shadow = blocktemp.getElementsByClassName('shadow')[0];\n",
       "\n",
       "      centerArc = blocktemp.getElementsByClassName('center-arc')[0];\n",
       "      blockLength = centerArc.getTotalLength();\n",
       "\n",
       "      nameLength = text.innerHTML.length;\n",
       "      while (1.2 * text.getComputedTextLength() > blockLength) {\n",
       "         nameLength -= 1;\n",
       "         // Trim until it isn't\n",
       "         text.innerHTML = text.innerHTML.slice(0, nameLength);\n",
       "         text.setAttribute('startOffset', '10%');\n",
       "         text.style['text-anchor'] = 'start';\n",
       "\n",
       "         shadow.innerHTML = shadow.innerHTML.slice(0, nameLength);\n",
       "         shadow.setAttribute('startOffset', '10%');\n",
       "         shadow.style['text-anchor'] = 'start';\n",
       "       }\n",
       "    }\n",
       "\n",
       "    // Add line features\n",
       "    var lineGroup = featuresGroup.append('g')\n",
       "        .attr('class', 'lineFeatureGroup');\n",
       "\n",
       "    var lineFeatures = lineGroup.selectAll('.feature')\n",
       "        .data(lineData)\n",
       "      .enter().append('g')\n",
       "      .append('path')\n",
       "        .attr('class', 'feature')\n",
       "        .attr('class', 'line-feature')\n",
       "        .attr('d', function(d, i) {\n",
       "          featureAngle = baseToRadians((d.start + d.stop) / 2);\n",
       "          inner = polarToCartesian(innerRadius - featureOffset - config.tickLength, featureAngle);\n",
       "          outer = polarToCartesian(outerRadius + featureOffset + config.tickLength, featureAngle);\n",
       "\n",
       "          return line([inner, outer]);\n",
       "        })\n",
       "        .attr('stroke', 'rgb(96, 96, 96)')\n",
       "        .attr('stroke-width', config.lineFeatureWidth)\n",
       "        .attr('fill', 'none')\n",
       "        .on('mouseover', function(d) {\n",
       "          angle = baseToRadians((d.start + d.stop) / 2);\n",
       "          position = polarToCartesian((outerRadius + featureOffset + config.tickLength + 3), angle);\n",
       "          lineGroup.append('text')\n",
       "              .attr('id', 'lineFeatureHighlight')\n",
       "              .attr('x', position.x)\n",
       "              .attr('y', position.y)\n",
       "              .attr('text-anchor', 'middle')\n",
       "              .attr('font-family', 'sans-serif')\n",
       "              .attr('font-size', '14px')\n",
       "              .style('font-weight', 'bold')\n",
       "              .attr('fill', 'red')\n",
       "              .attr('dy', '-1em')\n",
       "              .attr('transform', function(d) {\n",
       "                degrees = (angle + Math.PI) / (2 * Math.PI) * 360;\n",
       "                return 'rotate(' + (degrees - 90) + ' ' + (position.x) + ',' + (position.y) + ')';\n",
       "              })\n",
       "              .text(d.name);\n",
       "        })\n",
       "        .on('mouseout', function(d) {\n",
       "          d3.select('#lineFeatureHighlight').remove();\n",
       "        });\n",
       "    } else {\n",
       "\n",
       "    ///////////////////////////////////////\n",
       "    // Placement rules for linear sequences\n",
       "    ///////////////////////////////////////\n",
       "    var backboneHeight = 25;  // TODO: replace with unified thickness param\n",
       "    var xMin = config.padding;\n",
       "    var xMax = config.width - config.padding;\n",
       "\n",
       "    var backboneDim = {\n",
       "      'left': xMin,\n",
       "      'right': xMax,\n",
       "      'top': center.y - backboneHeight / 2,\n",
       "      'bottom': center.y + backboneHeight / 2\n",
       "    };\n",
       "\n",
       "    // Offset of 1 pixel makes the display look nicer\n",
       "    var xScale = d3.scale.linear()\n",
       "        .domain([0, sequence_json.sequence.length])\n",
       "        .range([xMin + 1, xMax + 1]);\n",
       "\n",
       "    ////////////////////\n",
       "    // Draw the backbone\n",
       "    ////////////////////\n",
       "    // TODO: similarities with circular - same fill, stroke, etc. Could simplify\n",
       "    //       with a backboneWidth attribute that's in pixels\n",
       "    svg.append('rect')\n",
       "        .attr('class', 'backbone')\n",
       "        .attr('x', backboneDim.left)\n",
       "        .attr('y', backboneDim.top)\n",
       "        .attr('width', backboneDim.right - backboneDim.left)\n",
       "        .attr('height', backboneDim.bottom - backboneDim.top)\n",
       "        .style('fill', config.backboneColor)\n",
       "        .style('stroke', d3.rgb(config.backboneColor).darker(0.5))\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    /////////////////////////////////\n",
       "    // Add plasmid name, number of bp\n",
       "    /////////////////////////////////\n",
       "    // TODO: similarities with circular - identical, just different placement\n",
       "    // FIXME: make the placement of text have no magic numbers\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.name)\n",
       "        .attr('x', center.x)\n",
       "        .attr('y', backboneDim.bottom +  50)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', config.titleFontSize);\n",
       "\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.sequence.length + ' bp')\n",
       "        .attr('x', center.x)\n",
       "        .attr('y', backboneDim.bottom + 80)\n",
       "        .style('text-anchor', 'middle');\n",
       "\n",
       "    /////////////\n",
       "    // Tick Marks\n",
       "    /////////////\n",
       "    // TODO: similarities with circular - almost everything\n",
       "    // Generate ticks data\n",
       "    var tickData = [];\n",
       "    for (var i=0; i < config.ticks; i++) {\n",
       "      tickData.push(i * Math.floor(sequence_json.sequence.length / config.ticks));\n",
       "    }\n",
       "\n",
       "    var tickGroup = svg.append('g')\n",
       "        .attr('id', 'tick-group');\n",
       "\n",
       "    var ticks = tickGroup.selectAll('.tick')\n",
       "        .data(tickData)\n",
       "      .enter().append('line')\n",
       "        .attr('x1', function(d) { return xScale(d); })\n",
       "        .attr('x2', function(d) { return xScale(d); })\n",
       "        .attr('y1', backboneDim.bottom)\n",
       "        .attr('y2', backboneDim.bottom + config.tickLength)\n",
       "        .attr('stroke', function(d, i) {\n",
       "          if (i % config.majorTickInterval === 0) {\n",
       "            tickColor = config.tickColorMajor;\n",
       "          } else {\n",
       "            tickColor = config.tickColorMinor;\n",
       "          }\n",
       "          return tickColor;\n",
       "        })\n",
       "        .attr('stroke-width', 2)\n",
       "        .attr('fill', 'none');\n",
       "\n",
       "    var tickLabels = tickGroup.selectAll('.tick-label')\n",
       "        .data(tickData)\n",
       "      .enter().append('text')\n",
       "        .text(function(d) {\n",
       "            if (d % config.majorTickInterval === 0) {\n",
       "              return d;\n",
       "            }\n",
       "        })\n",
       "        .attr('x', function(d) { return xScale(d); })\n",
       "        .attr('y', function(d) { return backboneDim.bottom + config.tickLength + 8; })\n",
       "        .attr('dy', '0.4em')\n",
       "        .attr('fill', config.tickColorMinor)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', '12px');\n",
       "\n",
       "    ////////////////////////////////\n",
       "    // Add features and their labels\n",
       "    ////////////////////////////////\n",
       "\n",
       "    // Separate out arc features and line features\n",
       "    var blockData = [];\n",
       "    var lineData = [];\n",
       "    // FIXME: this process is broken and should be determined by fit, not n\n",
       "    // arbitrary scaling factor\n",
       "    for (var i = 0; i < sequence_json.features.length; i++) {\n",
       "      feature = sequence_json.features[i];\n",
       "      if (xScale(feature.stop - feature.start) > 10 * config.minFeatureSize) {\n",
       "        blockData.push(feature);\n",
       "      } else {\n",
       "        lineData.push(feature);\n",
       "      }\n",
       "    }\n",
       "\n",
       "    var featuresGroup = svg.append('g')\n",
       "        .attr('id', 'features');\n",
       "    var blockGroup = featuresGroup.append('g')\n",
       "        .attr('id', 'block-features');\n",
       "\n",
       "    var blockFeatures = blockGroup.selectAll('.feature')\n",
       "        .data(blockData)\n",
       "      .enter().append('g')\n",
       "        .attr('class', 'feature')\n",
       "      .append('rect')\n",
       "        .attr('class', 'feature-block')\n",
       "        .attr('x', function(d) { return xScale(d.start); })\n",
       "        .attr('y', function(d) { return backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2; })\n",
       "        .attr('width', function(d) { return xScale(d.stop) - xScale(d.start); })\n",
       "        .attr('height', function(d) { return config.featureScale * backboneHeight; })\n",
       "        .attr('id', function(d, i) { return 'block' + i; })\n",
       "        .style('fill', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            return d.color;\n",
       "          } else {\n",
       "            return 'rgb(192, 192, 256)';\n",
       "          }\n",
       "        })\n",
       "        .style('fill-opacity', config.featureOpacity)\n",
       "        .style('stroke', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            color = d.color;\n",
       "          } else {\n",
       "            color = 'rgb(192, 192, 256)';\n",
       "          }\n",
       "          return d3.rgb(color).darker(0.5);\n",
       "        })\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    // Gaussian blur filter for making text shadow\n",
       "    var filter = svg.append('defs')\n",
       "      .append('filter')\n",
       "        .attr('id', 'text-shadow')\n",
       "        .attr('height', '130%')\n",
       "      .append('feGaussianBlur')\n",
       "        .attr('in', 'SourceAlpha')\n",
       "        .attr('stdDeviation', 2)\n",
       "        .attr('result', 'blur');\n",
       "\n",
       "    var shadows = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'shadow')\n",
       "        .attr('id', function(d, i) {\n",
       "          return 'shadow' + i;\n",
       "        })\n",
       "        .attr('x', function(d) { return xScale((d.start + d.stop) / 2); })\n",
       "        .attr('y', function(d) { return (backboneDim.top + backboneDim.bottom) / 2 + 3; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '10px')\n",
       "        .style('filter', 'url(#text-shadow)')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    var blockText = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-text')\n",
       "        .attr('fill', 'white')\n",
       "        .attr('id', function(d, i) { return 'label' + i; })\n",
       "        .attr('x', function(d) { return xScale((d.start + d.stop) / 2); })\n",
       "        .attr('y', function(d) { return (backboneDim.top + backboneDim.bottom) / 2; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '10px')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    // If text is too big, don't offset and set anchor to left\n",
       "    // FIXME: this should really be done with more d3-style methods\n",
       "    feature_selection = document.getElementsByClassName('feature');\n",
       "    for (var i = 0; i < feature_selection.length; i++) {\n",
       "      blocktemp = feature_selection[i];\n",
       "      text = blocktemp.getElementsByClassName('block-text')[0];\n",
       "      shadow = blocktemp.getElementsByClassName('shadow')[0];\n",
       "\n",
       "      block = blocktemp.getElementsByClassName('feature-block');\n",
       "      blockLength = block[0].width.baseVal.value\n",
       "\n",
       "      nameLength = text.innerHTML.length;\n",
       "      while (1.2 * text.getComputedTextLength() > blockLength) {\n",
       "         nameLength -= 1;\n",
       "         // Trim until it isn't\n",
       "         text.innerHTML = text.innerHTML.slice(0, nameLength);\n",
       "\n",
       "         shadow.innerHTML = shadow.innerHTML.slice(0, nameLength);\n",
       "       }\n",
       "    }\n",
       "\n",
       "    // Add line features\n",
       "    var lineGroup = featuresGroup.append('g')\n",
       "        .attr('class', 'lineFeatureGroup');\n",
       "\n",
       "    var lineFeatures = lineGroup.selectAll('.feature')\n",
       "        .data(lineData)\n",
       "      .enter().append('g')\n",
       "      .append('line')\n",
       "        .attr('class', 'feature')\n",
       "        .attr('class', 'line-feature')\n",
       "        .attr('x1', function(d) { return xScale(d.start); })\n",
       "        .attr('x2', function(d) { return xScale(d.start); })\n",
       "        .attr('y1', function(d) { return backboneDim.bottom + (config.featureScale * backboneHeight - backboneHeight) / 2 + config.tickLength; })\n",
       "        .attr('y2', function(d) { return backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2 - config.tickLength; })\n",
       "        .attr('stroke', 'rgb(96, 96, 96)')\n",
       "        .attr('stroke-width', config.lineFeatureWidth)\n",
       "        .attr('fill', 'none')\n",
       "        .on('mouseover', function(d) {\n",
       "          lineGroup.append('text')\n",
       "              .attr('id', 'lineFeatureHighlight')\n",
       "              .attr('x', xScale(d.start))\n",
       "              .attr('y', backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2 - 3)\n",
       "              .attr('text-anchor', 'middle')\n",
       "              .attr('font-family', 'sans-serif')\n",
       "              .attr('font-size', '14px')\n",
       "              .style('font-weight', 'bold')\n",
       "              .attr('fill', 'red')\n",
       "              .attr('dy', '-1em')\n",
       "              .text(d.name);\n",
       "        })\n",
       "        .on('mouseout', function(d) {\n",
       "          d3.select('#lineFeatureHighlight').remove();\n",
       "        });\n",
       "\n",
       "  }\n",
       "}\n",
       "\n",
       "            d3sequence(window.data, 'sequence_e09671b4-c47f-11e5-9eaa-02591d3bb8ad')\n",
       "        });\n",
       "        </script>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "backbone_source = cr.seqio.read_dna('./pmod6-pgpd-mcherry.gb')\n",
    "print type(backbone_source)\n",
    "print len(backbone_source)\n",
    "backbone_source.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id=sequence_e0d77bc8-c47f-11e5-9eaa-02591d3bb8ad></div>\n",
       "        <script>\n",
       "        require(['//d3js.org/d3.v3.min.js'], function(lib) {\n",
       "            window.data = {\"features\": [{\"color\": [\"#b7e6d7\"], \"start\": 4446, \"stop\": 4470, \"name\": \"PP2\", \"type\": \"primer_bind\"}, {\"color\": [\"#c7b0e3\"], \"start\": 4470, \"stop\": 5125, \"name\": \"Actual GPD Promoter\", \"type\": \"promoter\"}, {\"color\": [\"#f8d3a9\"], \"start\": 0, \"stop\": 22, \"name\": \"TP\", \"type\": \"misc_feature\"}, {\"color\": [\"#f58a5e\"], \"start\": 22, \"stop\": 262, \"name\": \"CYC1_terminator\", \"type\": \"terminator\"}, {\"color\": [\"#b1ff67\"], \"start\": 31, \"stop\": 50, \"name\": \"pYESTrp_rev_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#f8d3a9\"], \"start\": 31, \"stop\": 50, \"name\": \"CYC1_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#9eafd2\"], \"start\": 262, \"stop\": 284, \"name\": \"TS\", \"type\": \"primer_bind\"}, {\"color\": [\"#ffef86\"], \"start\": 284, \"stop\": 784, \"name\": \"URA3 3' UTR\", \"type\": \"3'UTR\"}, {\"color\": [\"#b7e6d7\"], \"start\": 784, \"stop\": 792, \"name\": \"PmeI(1)\", \"type\": \"misc_feature\"}, {\"color\": [\"#d6b295\"], \"start\": 796, \"stop\": 813, \"name\": \"M13 Reverse primer\", \"type\": \"promoter\"}, {\"color\": [\"#c6c9d1\"], \"start\": 809, \"stop\": 832, \"name\": \"M13_pUC_rev_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#f8d3a9\"], \"start\": 851, \"stop\": 852, \"name\": \"mutant?\", \"type\": \"misc_feature\"}, {\"color\": [\"#75c6a9\"], \"start\": 1183, \"stop\": 1803, \"name\": \"pBR322_origin\", \"type\": \"rep_origin\"}, {\"color\": [\"#85dae9\"], \"start\": 1211, \"stop\": 1212, \"name\": \"mut?\", \"type\": \"misc_feature\"}, {\"color\": [\"#75c6a9\"], \"start\": 1957, \"stop\": 2818, \"name\": \"Ampicillin\", \"type\": \"CDS\"}, {\"color\": [\"#f8d3a9\"], \"start\": 2859, \"stop\": 2888, \"name\": \"AmpR_promoter\", \"type\": \"promoter\"}, {\"color\": [\"#c7b0e3\"], \"start\": 3046, \"stop\": 3069, \"name\": \"pGEX_3_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#85dae9\"], \"start\": 3204, \"stop\": 3212, \"name\": \"PmeI\", \"type\": \"misc_feature\"}, {\"color\": [\"#b1ff67\"], \"start\": 3212, \"stop\": 3460, \"name\": \"URA3 promoter (kind of)\", \"type\": \"promoter\"}, {\"color\": [\"#b4abac\"], \"start\": 3460, \"stop\": 4165, \"name\": \"URA3\", \"type\": \"gene\"}, {\"color\": [\"#85dae9\"], \"start\": 4160, \"stop\": 4161, \"name\": \"mutated to A in ura3-1 strains\", \"type\": \"misc_difference\"}, {\"color\": [\"#faac61\"], \"start\": 4264, \"stop\": 4446, \"name\": \"ADH1 Terminator\", \"type\": \"terminator\"}], \"bottom\": \"TTTGTAGATCCACTAGTTCTAGAATCCGATCCGTCGAAACTAAGTTCTGGTGTTTTAAAACTAAAAAAAAGACTAACTATAAAAGTAGAATTTAAGAAGTTTAAGAAATAGATTTACAGAATTACAATCAATACCTACCGTCTTTATATACTTATTAGTCAAGTAGGGGAATAATTTCAGGGAACTGGTTTCAACCTTTTTTTTCAGCTTTTTCCAAATCAGAGAGAGCAGAAGGTAATAGAAGGTGTAAGAAAATGAGATAGATACATGCGTGGGTCAATTGCCTTGTGTCATCATTTACTCCAGGCAGGTTGCATCACTCCATTGAGGTTGTGCCCGTTTTTTGCCTGTTTGTGCCCCTGTTCTCTGTAGTTGCGCTAAGAGAATGGACCTATGAACTGATGGTTGGTGAAGAAAACAATATTTTGGTGCTGGGATTCTTTTTTTTTCTGGATGCCAGCTTAAAAAGCGGGCTCCATTATATTTAGTGGATGCCAGGAATAAACTGTTCACCCAGACACCTACGATGTTATATATTCTGTGTAACCCGCCCCCTATTTTGGGCATGTACGGGTTACAGCAGAATTAAAAGGCTAATTTTTTGACTAAATAAAGTTAGGAAAATCACTACTATTAATTATTTACGTATTCTTTGAAATGGCGAGTATTGATAATGATAAACTGCCGCTTTTCGAATCTGGTTAAGGGAGGTGTGGTCAATAAGAGCGACCTCATGCTATACCTGAGAAAGCAACCTGACCTACAGGAAAGAGTTACTCAAGAATAAGAATTTTCGTTTTAAAACCTAAGAGTCACTTTAAAATTTGTATACACTTATTTTTTTTATAACTTATTTAATAATAAAAATCATAAATCATAAGAAATTCGCTTAGTTTTGCTGGCCGCATCTTCTCAAATATGCTTCCCAGCCTGCTTTTCTGTAACGTTCACCCTCTACCTTAGCATCCCTTCCCTTTGCAAATAGTCCTCTTCCAACAATAATAATGTCAGATCCTGTAGAGACCACATCATCCACGGTTCTATACTGTTGACCCAATGCGTCTCCCTTGTCATCTAAACCCACACCGGGTGTCATAATCAACCAATCGTAACCTTCATCTCTTCCACCCATGTCTCTTTGAGCAATAAAGCCGATAACAAAATCTTTGTCGCTCTTCGCAATGTCAACAGTACCCTTAGTATATTCTCCAGTAGCTAGGGAGCCCTTGCATGACAATTCTGCTAACATCAAAAGGCCTCTAGGTTCCTTTGTTACTTCTTCCGCCGCCTGCTTCAAACCGCTAACAATACCTGGGCCCACCACACCGTGTGCATTCGTAATGTCTGCCCATTCTGCTATTCTGTATACACCCGCAGAGTACTGCAATTTGACTGTATTACCAATGTCAGCAAATTTTCTGTCTTCGAAGAGTAAAAAATTGTACTTGGCGGATAATGCCTTTAGCGGCTTAACTGTGCCCTCCATGGAAAAATCAGTCAAGATATCCACATGTGTTTTTAGTAAACAAATTTTGGGACCTAATGCTTCAACTAACTCCAGTAATTCCTTGGTGGTACGAACATCCAATGAAGCACACAAGTTTGTTTGCTTTTCGTGCATGATATTAAATAGCTTGGCAGCAACAGGACTAGGATGAGTAGCAGCACGTTCCTTATATGTAGCTTTCGACATGATTTATCTTCGTTTCCTGCAGGTTTTTGTTCTGTGCAGTTGGGTTAAGAATACTGGGCAATTTCATGTTTCTTCAACACCACATATGCGTATATATACCAATCTAAGTCTGTGCTCCTTCCTTCGTTCTTCCTTCTGCTCGGAGATTACCGAATCAAAAAAATTTCAAAGAAACCGGAATCAAAAAAAAGAACAAAAAAAAAAAAGATGAATTGAAAAGCTTTATGGACCCTGAAACCACAGCCACAGTTTAAACTATGGTGCACTCTCAGTACAATCTGCTCTGATGCCGCATAGTTAAGCCAGCCCCGACACCCGCCAACACCCGCTGACGCGCCCTGACGGGCTTGTCTGCTCCCGGCATCCGCTTACAGACAAGCTGTGACCGTCTCCGGGAGCTGCATGTGTCAGAGGTTTTCACCGTCATCACCGAAACGCGCGAGACGAAAGGGCCTCGTGATACGCCTATTTTTATAGGTTAATGTCATGATAATAATGGTTTCTTAGACGTCAGGTGGCACTTTTCGGGGAAATGTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATTCAAATATGTATCCGCTCATGAGACAATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAGTATGAGTATTCAACATTTCCGTGTCGCCCTTATTCCCTTTTTTGCGGCATTTTGCCTTCCTGTTTTTGCTCACCCAGAAACGCTGGTGAAAGTAAAAGATGCTGAAGATCAGTTGGGTGCACGAGTGGGTTACATCGAACTGGATCTCAACAGCGGTAAGATCCTTGAGAGTTTTCGCCCCGAAGAACGTTTTCCAATGATGAGCACTTTTAAAGTTCTGCTATGTGGCGCGGTATTATCCCGTATTGACGCCGGGCAAGAGCAACTCGGTCGCCGCATACACTATTCTCAGAATGACTTGGTTGAGTACTCACCAGTCACAGAAAAGCATCTTACGGATGGCATGACAGTAAGAGAATTATGCAGTGCTGCCATAACCATGAGTGATAACACTGCGGCCAACTTACTTCTGACAACGATCGGAGGACCGAAGGAGCTAACCGCTTTTTTTCACAACATGGGGGATCATGTAACTCGCCTTGATCGTTGGGAACCGGAGCTGAATGAAGCCATACCAAACGACGAGCGTGACACCACGATGCCTGTAGCAATGGCAACAACGTTGCGCAAACTATTAACTGGCGAACTACTTACTCTAGCTTCCCGGCAACAATTAATAGACTGGATGGAGGCGGATAAAGTTGCAGGACCACTTCTGCGCTCGGCCCTTCCGGCTGGCTGGTTTATTGCTGATAAATCTGGAGCCGGTGAGCGTGGGTCTCGCGGTATCATTGCAGCACTGGGGCCAGATGGTAAGCCCTCCCGTATCGTAGTTATCTACACGACGGGCAGTCAGGCAACTATGGATGAACGAAATAGACAGATCGCTGAGATAGGTGCCTCACTGATTAAGCATTGGTAACTGTCAGACCAAGTTTACTCATATATACTTTAGATTGATTTAAAACTTCATTTTTAATTTAAAAGGATCTAGGTGAAGATCCTTTTTGATAATCTCATGACCAAAATCCCTTAACGTGAGTTTTCGTTCCACTGAGCGTCAGACCCCGTAGAAAAGATCAAAGGATCTTCTTGAGATCCTTTTTTTCTGCGCGTAATCTGCTGCTTGCAAACAAAAAAACCACCGCTACCAGCGGTGGTTTGTTTGCCGGATCAAGAGCTACCAACTCTTTTTCCGAAGGTAACTGGCTTCAGCAGAGCGCAGATACCAAATACTGTCCTTCTAGTGTAGCCGTAGTTAGGCCACCACTTCAAGAACTCTGTAGCACCGCCTACATACCTCGCTCTGCTAATCCTGTTACCAGTGGCTGCTGCCAGTGGCGATAAGTCGTGTCTTACCGGGTTGGACTCAAGACGATAGTTACCGGATAAGGCGCAGCGGTCGGGCTGAACGGGGGGTTCGTGCACACAGCCCAGCTTGGAGCGAACGACCTACACCGAACTGAGATACCTACAGCGTGAGCATTGAGAAAGCGCCACGCTTCCCGAAGGGAGAAAGGCGGACAGGTATCCGGTAAGCGGCAGGGTCGGAACAGGAGAGCGCACGAGGGAGCTTCCAGGGGGGAACGCCTGGTATCTTTATAGTCCTGTCGGGTTTCGCCACCTCTGACTTGAGCGTCGATTTTTGTGATGCTCGTCAGGGGGGCGGAGCCTATGGAAAAACGCCAGCAACGCGGCCTTTTTACGGTTCCTGGCCTTTTGCTGGCCTTTTGCTCACATGTTCTTTCCTGCGTTATCCCCTGATTCTGTGGATAACCGTATTACCGCCTTTGAGTGAGCTGATACCGCTCGCCGCAGCCGAACGACCGAGCGCAGCGAGTCAGTGAGCGAGGAAGCGGAAGAGCGCCCAATACGCAAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTGAGCGCAACGCAATTAATGTGAGTTACCTCACTCATTAGGCACCCCAGGCTTTACACTTTATGCTTCCGGCTCGTATGTTGTGTGGAATTGTGAGCGGATAACAATTTCACACAGGAAACAGCTATGACCATGGTTTAAACCGGCCCAAGCCTTGTCCCAAGGCAGCGTTTTGTTCTTGGAAACGCTGCCCTACACGTTCGCTATGCTTCAAGAACTTTTCTGAGCACTTCATGATGCATGTTTGTTCCTTATTGGTTAGCTTTGATGTTGTGAAGTCATTGACACAGTCTGTGAAACATCTTTCTACCAGATTAGAGTACAAACGCATGAAATCCTTCATTTGCTTTTGTTCCACTACTTTTTGGAACTCTTGTTGTTCTTTGGAGTTCAATGCGTCCATCTTTACAGTCCTGTCTTATTGTTCTTGATTTGTGCCCCGTAAAATACTGTTACTTGGTTCTGGCGAGGTATTGGATAGTTCCTTTTTATAAAGGCCATGAAGCTTTTTCTTTCCAATTTTTTTTTTTTCGTCATTATAGAAATCATTACGACCGAGATTCCCGGGTAATAACTGATATAATTAAATTGAAGCTCTAATTTGTGAGTTTAGTATACATGCATTTACTTATAATACAGTTTTCGGGGCGGGGTTATTACGACATCAAATTAAAGCCTTCGAGCGTCCCAAAACCTTCTCAAGCAAGGTTTTCAGTATAATGTTACATGCGTACACGCGTCTGTACAGAAAAAAAAGAAAAATTTGAAATATAAATAACGTTCTTAATACTAACATAACTATAAAAAAATAAATAGGGACCTAGACTTCAGGTTGTCTAACTCCTTCCTTTTCGGTTAGAGCGGATGTGGGGGGAGGGCGTGAATGTAAGCGTGACATAACTAATTGACTCGAGGTCGACGGTATCA\", \"material\": \"DNA\", \"sequence\": \"TGATACCGTCGACCTCGAGTCAATTAGTTATGTCACGCTTACATTCACGCCCTCCCCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAGACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGATGTCGTAATAACCCCGCCCCGAAAACTGTATTATAAGTAAATGCATGTATACTAAACTCACAAATTAGAGCTTCAATTTAATTATATCAGTTATTACCCGGGAATCTCGGTCGTAATGATTTCTATAATGACGAAAAAAAAAAAATTGGAAAGAAAAAGCTTCATGGCCTTTATAAAAAGGAACTATCCAATACCTCGCCAGAACCAAGTAACAGTATTTTACGGGGCACAAATCAAGAACAATAAGACAGGACTGTAAAGATGGACGCATTGAACTCCAAAGAACAACAAGAGTTCCAAAAAGTAGTGGAACAAAAGCAAATGAAGGATTTCATGCGTTTGTACTCTAATCTGGTAGAAAGATGTTTCACAGACTGTGTCAATGACTTCACAACATCAAAGCTAACCAATAAGGAACAAACATGCATCATGAAGTGCTCAGAAAAGTTCTTGAAGCATAGCGAACGTGTAGGGCAGCGTTTCCAAGAACAAAACGCTGCCTTGGGACAAGGCTTGGGCCGGTTTAAACCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAATTCCACACAACATACGAGCCGGAAGCATAAAGTGTAAAGCCTGGGGTGCCTAATGAGTGAGGTAACTCACATTAATTGCGTTGCGCTCACTGCCCGCTTTCCAGTCGGGAAACCTGTCGTGCCAGCTGCATTAATGAATCGGCCAACGCGCGGGGAGAGGCGGTTTGCGTATTGGGCGCTCTTCCGCTTCCTCGCTCACTGACTCGCTGCGCTCGGTCGTTCGGCTGCGGCGAGCGGTATCAGCTCACTCAAAGGCGGTAATACGGTTATCCACAGAATCAGGGGATAACGCAGGAAAGAACATGTGAGCAAAAGGCCAGCAAAAGGCCAGGAACCGTAAAAAGGCCGCGTTGCTGGCGTTTTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAAAATCGACGCTCAAGTCAGAGGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTCCCCCCTGGAAGCTCCCTCGTGCGCTCTCCTGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGGAAGCGTGGCGCTTTCTCAATGCTCACGCTGTAGGTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTGGGCTGTGTGCACGAACCCCCCGTTCAGCCCGACCGCTGCGCCTTATCCGGTAACTATCGTCTTGAGTCCAACCCGGTAAGACACGACTTATCGCCACTGGCAGCAGCCACTGGTAACAGGATTAGCAGAGCGAGGTATGTAGGCGGTGCTACAGAGTTCTTGAAGTGGTGGCCTAACTACGGCTACACTAGAAGGACAGTATTTGGTATCTGCGCTCTGCTGAAGCCAGTTACCTTCGGAAAAAGAGTTGGTAGCTCTTGATCCGGCAAACAAACCACCGCTGGTAGCGGTGGTTTTTTTGTTTGCAAGCAGCAGATTACGCGCAGAAAAAAAGGATCTCAAGAAGATCCTTTGATCTTTTCTACGGGGTCTGACGCTCAGTGGAACGAAAACTCACGTTAAGGGATTTTGGTCATGAGATTATCAAAAAGGATCTTCACCTAGATCCTTTTAAATTAAAAATGAAGTTTTAAATCAATCTAAAGTATATATGAGTAAACTTGGTCTGACAGTTACCAATGCTTAATCAGTGAGGCACCTATCTCAGCGATCTGTCTATTTCGTTCATCCATAGTTGCCTGACTGCCCGTCGTGTAGATAACTACGATACGGGAGGGCTTACCATCTGGCCCCAGTGCTGCAATGATACCGCGAGACCCACGCTCACCGGCTCCAGATTTATCAGCAATAAACCAGCCAGCCGGAAGGGCCGAGCGCAGAAGTGGTCCTGCAACTTTATCCGCCTCCATCCAGTCTATTAATTGTTGCCGGGAAGCTAGAGTAAGTAGTTCGCCAGTTAATAGTTTGCGCAACGTTGTTGCCATTGCTACAGGCATCGTGGTGTCACGCTCGTCGTTTGGTATGGCTTCATTCAGCTCCGGTTCCCAACGATCAAGGCGAGTTACATGATCCCCCATGTTGTGAAAAAAAGCGGTTAGCTCCTTCGGTCCTCCGATCGTTGTCAGAAGTAAGTTGGCCGCAGTGTTATCACTCATGGTTATGGCAGCACTGCATAATTCTCTTACTGTCATGCCATCCGTAAGATGCTTTTCTGTGACTGGTGAGTACTCAACCAAGTCATTCTGAGAATAGTGTATGCGGCGACCGAGTTGCTCTTGCCCGGCGTCAATACGGGATAATACCGCGCCACATAGCAGAACTTTAAAAGTGCTCATCATTGGAAAACGTTCTTCGGGGCGAAAACTCTCAAGGATCTTACCGCTGTTGAGATCCAGTTCGATGTAACCCACTCGTGCACCCAACTGATCTTCAGCATCTTTTACTTTCACCAGCGTTTCTGGGTGAGCAAAAACAGGAAGGCAAAATGCCGCAAAAAAGGGAATAAGGGCGACACGGAAATGTTGAATACTCATACTCTTCCTTTTTCAATATTATTGAAGCATTTATCAGGGTTATTGTCTCATGAGCGGATACATATTTGAATGTATTTAGAAAAATAAACAAATAGGGGTTCCGCGCACATTTCCCCGAAAAGTGCCACCTGACGTCTAAGAAACCATTATTATCATGACATTAACCTATAAAAATAGGCGTATCACGAGGCCCTTTCGTCTCGCGCGTTTCGGTGATGACGGTGAAAACCTCTGACACATGCAGCTCCCGGAGACGGTCACAGCTTGTCTGTAAGCGGATGCCGGGAGCAGACAAGCCCGTCAGGGCGCGTCAGCGGGTGTTGGCGGGTGTCGGGGCTGGCTTAACTATGCGGCATCAGAGCAGATTGTACTGAGAGTGCACCATAGTTTAAACTGTGGCTGTGGTTTCAGGGTCCATAAAGCTTTTCAATTCATCTTTTTTTTTTTTGTTCTTTTTTTTGATTCCGGTTTCTTTGAAATTTTTTTGATTCGGTAATCTCCGAGCAGAAGGAAGAACGAAGGAAGGAGCACAGACTTAGATTGGTATATATACGCATATGTGGTGTTGAAGAAACATGAAATTGCCCAGTATTCTTAACCCAACTGCACAGAACAAAAACCTGCAGGAAACGAAGATAAATCATGTCGAAAGCTACATATAAGGAACGTGCTGCTACTCATCCTAGTCCTGTTGCTGCCAAGCTATTTAATATCATGCACGAAAAGCAAACAAACTTGTGTGCTTCATTGGATGTTCGTACCACCAAGGAATTACTGGAGTTAGTTGAAGCATTAGGTCCCAAAATTTGTTTACTAAAAACACATGTGGATATCTTGACTGATTTTTCCATGGAGGGCACAGTTAAGCCGCTAAAGGCATTATCCGCCAAGTACAATTTTTTACTCTTCGAAGACAGAAAATTTGCTGACATTGGTAATACAGTCAAATTGCAGTACTCTGCGGGTGTATACAGAATAGCAGAATGGGCAGACATTACGAATGCACACGGTGTGGTGGGCCCAGGTATTGTTAGCGGTTTGAAGCAGGCGGCGGAAGAAGTAACAAAGGAACCTAGAGGCCTTTTGATGTTAGCAGAATTGTCATGCAAGGGCTCCCTAGCTACTGGAGAATATACTAAGGGTACTGTTGACATTGCGAAGAGCGACAAAGATTTTGTTATCGGCTTTATTGCTCAAAGAGACATGGGTGGAAGAGATGAAGGTTACGATTGGTTGATTATGACACCCGGTGTGGGTTTAGATGACAAGGGAGACGCATTGGGTCAACAGTATAGAACCGTGGATGATGTGGTCTCTACAGGATCTGACATTATTATTGTTGGAAGAGGACTATTTGCAAAGGGAAGGGATGCTAAGGTAGAGGGTGAACGTTACAGAAAAGCAGGCTGGGAAGCATATTTGAGAAGATGCGGCCAGCAAAACTAAGCGAATTTCTTATGATTTATGATTTTTATTATTAAATAAGTTATAAAAAAAATAAGTGTATACAAATTTTAAAGTGACTCTTAGGTTTTAAAACGAAAATTCTTATTCTTGAGTAACTCTTTCCTGTAGGTCAGGTTGCTTTCTCAGGTATAGCATGAGGTCGCTCTTATTGACCACACCTCCCTTAACCAGATTCGAAAAGCGGCAGTTTATCATTATCAATACTCGCCATTTCAAAGAATACGTAAATAATTAATAGTAGTGATTTTCCTAACTTTATTTAGTCAAAAAATTAGCCTTTTAATTCTGCTGTAACCCGTACATGCCCAAAATAGGGGGCGGGTTACACAGAATATATAACATCGTAGGTGTCTGGGTGAACAGTTTATTCCTGGCATCCACTAAATATAATGGAGCCCGCTTTTTAAGCTGGCATCCAGAAAAAAAAAGAATCCCAGCACCAAAATATTGTTTTCTTCACCAACCATCAGTTCATAGGTCCATTCTCTTAGCGCAACTACAGAGAACAGGGGCACAAACAGGCAAAAAACGGGCACAACCTCAATGGAGTGATGCAACCTGCCTGGAGTAAATGATGACACAAGGCAATTGACCCACGCATGTATCTATCTCATTTTCTTACACCTTCTATTACCTTCTGCTCTCTCTGATTTGGAAAAAGCTGAAAAAAAAGGTTGAAACCAGTTCCCTGAAATTATTCCCCTACTTGACTAATAAGTATATAAAGACGGTAGGTATTGATTGTAATTCTGTAAATCTATTTCTTAAACTTCTTAAATTCTACTTTTATAGTTAGTCTTTTTTTTAGTTTTAAAACACCAGAACTTAGTTTCGACGGATCGGATTCTAGAACTAGTGGATCTACAAA\", \"circular\": false, \"name\": \"\"};// Full JSON description of plasmid - in the future, this would be\n",
       "// downloaded dynamically\n",
       "function d3sequence(sequence_json, div_id) {\n",
       "  // TODO: Have methods for drawing circular vs. linear, keep core logic the same\n",
       "  // Option that might make it into a configurable version\n",
       "  // TODO: control more of these with CSS\n",
       "\n",
       "  var circularConfig = {\n",
       "    'width': 500,\n",
       "    'height': 500,\n",
       "    'padding': 75,\n",
       "    'featureScale': 1.2,\n",
       "    'featureOpacity': 0.5,\n",
       "    'backboneThickness': 0.2,\n",
       "    'backboneColor': '#eee',\n",
       "    'minFeatureSize': 5,\n",
       "    'ticks': 80,\n",
       "    'majorTickInterval': 5,\n",
       "    'tickColorMinor': '#94aab0',\n",
       "    'tickColorMajor': 'black',\n",
       "    'tickLength': 6,\n",
       "    'lineFeatureWidth': 4,\n",
       "    'titleFontSize': '25px'\n",
       "  };\n",
       "\n",
       "  var linearConfig = {\n",
       "    'width': 800,\n",
       "    'height': 150,\n",
       "    'padding': 35,\n",
       "    'featureScale': 1.2,\n",
       "    'featureOpacity': 0.5,\n",
       "    'backboneThickness': 0.2,\n",
       "    'backboneColor': '#eee',\n",
       "    'minFeatureSize': 5,\n",
       "    'ticks': 40,\n",
       "    'majorTickInterval': 5,\n",
       "    'tickColorMinor': '#94aab0',\n",
       "    'tickColorMajor': 'black',\n",
       "    'tickLength': 6,\n",
       "    'lineFeatureWidth': 4,\n",
       "    'titleFontSize': '25px'\n",
       "  };\n",
       "\n",
       "  if (!sequence_json.circular) {\n",
       "    config = linearConfig;\n",
       "  } else if (sequence_json.circular) {\n",
       "    config = circularConfig;\n",
       "  }\n",
       "\n",
       "  ///////////////////////\n",
       "  // Initialize svg space\n",
       "  ///////////////////////\n",
       "  var svg = d3.select('#' + div_id)\n",
       "    .append('svg')\n",
       "      .attr('width', config.width)\n",
       "      .attr('height', config.height)\n",
       "    .append('g');\n",
       "\n",
       "  var center = {'x': config.width / 2, 'y': config.height / 2};\n",
       "\n",
       "  if (sequence_json.circular) {\n",
       "\n",
       "    var outerRadius = Math.min(config.width, config.height) / 2 - config.padding;\n",
       "    var innerRadius = outerRadius - config.backboneThickness * outerRadius;\n",
       "    var centerRadius = d3.mean([outerRadius, innerRadius]);\n",
       "    var featureOffset = (config.featureScale * (outerRadius - innerRadius) - (outerRadius - innerRadius)) / 2;\n",
       "\n",
       "    /////////////////////////////////////////\n",
       "    // Placement rules for circular sequences\n",
       "    /////////////////////////////////////////\n",
       "    var centerTranslation = 'translate(' + center.x + ',' + center.y + ')';\n",
       "    svg.attr('transform', centerTranslation);\n",
       "\n",
       "    var baseToRadians = d3.scale.linear()\n",
       "        .domain([0, sequence_json.sequence.length])\n",
       "        .range([-(Math.PI / 2), 1.5 * Math.PI]);\n",
       "\n",
       "    // Placement around the plasmid is always prior to centering in the svg\n",
       "    function polarToCartesian(radius, angle) {\n",
       "      return {'x': radius * Math.cos(angle), 'y': radius * Math.sin(angle)};\n",
       "    }\n",
       "\n",
       "\n",
       "    ///////////////////////////////////////////\n",
       "    // Make an annulus for the plasmid backbone\n",
       "    ///////////////////////////////////////////\n",
       "\n",
       "    var annulus = d3.svg.arc()\n",
       "        .innerRadius(innerRadius)\n",
       "        .outerRadius(outerRadius)\n",
       "        .startAngle(0)\n",
       "        .endAngle(2 * Math.PI);\n",
       "\n",
       "    backbone = svg.append('path')\n",
       "        .attr('class', 'backbone')\n",
       "        .attr('d', annulus)\n",
       "        .style('fill-rule', 'evenodd')\n",
       "        .style('fill', config.backboneColor)\n",
       "        .style('stroke', d3.rgb(config.backboneColor).darker(0.5))\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    /////////////////////////////////\n",
       "    // Add plasmid name, number of bp\n",
       "    /////////////////////////////////\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.name)\n",
       "        .attr('x', 0)\n",
       "        .attr('y', 0)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', config.titleFontSize);\n",
       "\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.sequence.length + ' bp')\n",
       "        .attr('x', 0)\n",
       "        .attr('y', 40)\n",
       "        .style('text-anchor', 'middle');\n",
       "\n",
       "    /////////////\n",
       "    // Tick Marks\n",
       "    /////////////\n",
       "\n",
       "    // Generate ticks data\n",
       "    var tickData = [];\n",
       "    for (var i=0; i < config.ticks; i++) {\n",
       "      tickData.push(i * Math.floor(sequence_json.sequence.length / config.ticks));\n",
       "    }\n",
       "\n",
       "    // FIXME: why use paths to make Ticks when lines can be used?\n",
       "    var line = d3.svg.line()\n",
       "        .x(function(d) { return d.x; } )\n",
       "        .y(function(d) { return d.y; } );\n",
       "\n",
       "    tickGroup = svg.append('g')\n",
       "        .attr('id', 'tick-group');\n",
       "\n",
       "    ticks = tickGroup.selectAll('.tick')\n",
       "        .data(tickData)\n",
       "      .enter().append('path')\n",
       "        .attr('d', function(d, i) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          inner = polarToCartesian(outerRadius, tickAngle);\n",
       "          outer = polarToCartesian(outerRadius + config.tickLength, tickAngle);\n",
       "          return line([inner, outer]);\n",
       "        })\n",
       "        .attr('stroke', function(d, i) {\n",
       "          if (i === 0) {\n",
       "            tickColor = 'red';\n",
       "          } else if (i % config.majorTickInterval === 0) {\n",
       "            tickColor = config.tickColorMajor;\n",
       "          } else {\n",
       "            tickColor = config.tickColorMinor;\n",
       "          }\n",
       "          return tickColor;\n",
       "        })\n",
       "        .attr('stroke-width', 2)\n",
       "        .attr('fill', 'none');\n",
       "\n",
       "    tickLabels = tickGroup.selectAll('.tick-label')\n",
       "        .data(tickData)\n",
       "      .enter().append('text')\n",
       "        .text(function(d) {\n",
       "            if (d % config.majorTickInterval === 0) {\n",
       "              return d;\n",
       "            }\n",
       "        })\n",
       "        .attr('x', function(d) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          return polarToCartesian(outerRadius + 22, tickAngle).x;\n",
       "        })\n",
       "        .attr('y', function(d) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          return polarToCartesian(outerRadius + 22, tickAngle).y;\n",
       "        })\n",
       "        .attr('dy', '0.4em')\n",
       "        .attr('fill', config.tickColorMinor)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', '12px');\n",
       "\n",
       "    ////////////////////////////////\n",
       "    // Add features and their labels\n",
       "    ////////////////////////////////\n",
       "\n",
       "    // Separate out arc features and line features\n",
       "    var blockData = [];\n",
       "    var lineData = [];\n",
       "    for (var i = 0; i < sequence_json.features.length; i++) {\n",
       "      feature = sequence_json.features[i];\n",
       "      arcAngle = 2 * Math.PI / sequence_json.sequence.length * (feature.stop - feature.start);\n",
       "      arcLength = centerRadius * arcAngle;\n",
       "      if (arcLength > config.minFeatureSize) {\n",
       "        blockData.push(feature);\n",
       "      } else {\n",
       "        lineData.push(feature);\n",
       "      }\n",
       "    }\n",
       "\n",
       "    var featuresGroup = svg.append('g')\n",
       "        .attr('id', 'features');\n",
       "    var blockGroup = featuresGroup.append('g')\n",
       "        .attr('id', 'block-features');\n",
       "\n",
       "    var blockFeatures = blockGroup.selectAll('.feature')\n",
       "        .data(blockData)\n",
       "      .enter()\n",
       "        .append('g')\n",
       "        .attr('class', 'feature')\n",
       "        .append('path')\n",
       "        .attr('d', function(d, i) {\n",
       "          the_arc = d3.svg.arc()\n",
       "            // d3 svg arc origin is at the top\n",
       "            .startAngle(baseToRadians(d.start) + Math.PI / 2)\n",
       "            .endAngle(baseToRadians(d.stop) + Math.PI / 2)\n",
       "            .innerRadius(innerRadius - featureOffset)\n",
       "            .outerRadius(outerRadius + featureOffset)\n",
       "            (d, i);\n",
       "          return the_arc;\n",
       "        })\n",
       "        .attr('id', function(d, i) { return 'block' + i; })\n",
       "        .style('fill-rule', 'evenodd')\n",
       "        .style('fill', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            return d.color;\n",
       "          } else {\n",
       "            return 'rgb(192, 192, 256)';\n",
       "          }\n",
       "        })\n",
       "        .style('fill-opacity', config.featureOpacity)\n",
       "        .style('stroke', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            color = d.color;\n",
       "          } else {\n",
       "            color = 'rgb(192, 192, 256)';\n",
       "          }\n",
       "          return d3.rgb(color).darker(0.5);\n",
       "        })\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "\n",
       "    // Gaussian blur filter for making text shadow\n",
       "    var filter = svg.append('defs')\n",
       "      .append('filter')\n",
       "        .attr('id', 'text-shadow')\n",
       "        .attr('height', '130%')\n",
       "      .append('feGaussianBlur')\n",
       "        .attr('in', 'SourceAlpha')\n",
       "        .attr('stdDeviation', 2)\n",
       "        .attr('result', 'blur');\n",
       "\n",
       "    // Add invisible arcs for aligning text\n",
       "    function describeArc(radius, start, end) {\n",
       "      start_xy = polarToCartesian(radius, baseToRadians(start));\n",
       "      end_xy = polarToCartesian(radius, baseToRadians(end));\n",
       "\n",
       "      var d = [\n",
       "        'M', start_xy.x, start_xy.y,\n",
       "        'A', radius, radius, 0, 0, 1, end_xy.x, end_xy.y\n",
       "      ].join(' ');\n",
       "\n",
       "      return d;\n",
       "    };\n",
       "\n",
       "    var centerArcs = blockGroup.selectAll('.feature, .center-arcs')\n",
       "      .append('path')\n",
       "        .attr('d', function(d) {\n",
       "          return describeArc(0.97 * centerRadius, d.start, d.stop)\n",
       "        })\n",
       "        .attr('class', 'center-arc')\n",
       "        .attr('id', function(d, i) { return div_id + '-center-arc' + i; })\n",
       "        .style('stroke', 'none')\n",
       "        .style('fill', 'none');\n",
       "\n",
       "    var shadows = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-label-shadow')\n",
       "      .append('textPath')\n",
       "        .attr('class', 'shadow')\n",
       "        .attr('startOffset', '50%')\n",
       "        .attr('xlink:href', function(d, i) {\n",
       "          return '#' + div_id + '-center-arc' + i;\n",
       "        })\n",
       "        .attr('id', function(d, i) {\n",
       "          return 'shadow' + i;\n",
       "        })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '11px')\n",
       "        .style('filter', 'url(#text-shadow)')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    var blockText = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-label')\n",
       "      .append('textPath')\n",
       "        .attr('class', 'block-text')\n",
       "        .attr('fill', 'white')\n",
       "        .attr('startOffset', '50%')\n",
       "        .attr('xlink:href', function(d, i) {\n",
       "          return '#' + div_id + '-center-arc' + i;\n",
       "        })\n",
       "        .attr('id', function(d, i) { return 'label' + i; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '11px')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    // If text is too big, don't offset and set anchor to left\n",
       "    // FIXME: this should really be done with more d3-style methods\n",
       "    feature_selection = document.getElementsByClassName('feature');\n",
       "    for (var i = 0; i < feature_selection.length; i++) {\n",
       "      blocktemp = feature_selection[i];\n",
       "      text = blocktemp.getElementsByClassName('block-text')[0];\n",
       "      shadow = blocktemp.getElementsByClassName('shadow')[0];\n",
       "\n",
       "      centerArc = blocktemp.getElementsByClassName('center-arc')[0];\n",
       "      blockLength = centerArc.getTotalLength();\n",
       "\n",
       "      nameLength = text.innerHTML.length;\n",
       "      while (1.2 * text.getComputedTextLength() > blockLength) {\n",
       "         nameLength -= 1;\n",
       "         // Trim until it isn't\n",
       "         text.innerHTML = text.innerHTML.slice(0, nameLength);\n",
       "         text.setAttribute('startOffset', '10%');\n",
       "         text.style['text-anchor'] = 'start';\n",
       "\n",
       "         shadow.innerHTML = shadow.innerHTML.slice(0, nameLength);\n",
       "         shadow.setAttribute('startOffset', '10%');\n",
       "         shadow.style['text-anchor'] = 'start';\n",
       "       }\n",
       "    }\n",
       "\n",
       "    // Add line features\n",
       "    var lineGroup = featuresGroup.append('g')\n",
       "        .attr('class', 'lineFeatureGroup');\n",
       "\n",
       "    var lineFeatures = lineGroup.selectAll('.feature')\n",
       "        .data(lineData)\n",
       "      .enter().append('g')\n",
       "      .append('path')\n",
       "        .attr('class', 'feature')\n",
       "        .attr('class', 'line-feature')\n",
       "        .attr('d', function(d, i) {\n",
       "          featureAngle = baseToRadians((d.start + d.stop) / 2);\n",
       "          inner = polarToCartesian(innerRadius - featureOffset - config.tickLength, featureAngle);\n",
       "          outer = polarToCartesian(outerRadius + featureOffset + config.tickLength, featureAngle);\n",
       "\n",
       "          return line([inner, outer]);\n",
       "        })\n",
       "        .attr('stroke', 'rgb(96, 96, 96)')\n",
       "        .attr('stroke-width', config.lineFeatureWidth)\n",
       "        .attr('fill', 'none')\n",
       "        .on('mouseover', function(d) {\n",
       "          angle = baseToRadians((d.start + d.stop) / 2);\n",
       "          position = polarToCartesian((outerRadius + featureOffset + config.tickLength + 3), angle);\n",
       "          lineGroup.append('text')\n",
       "              .attr('id', 'lineFeatureHighlight')\n",
       "              .attr('x', position.x)\n",
       "              .attr('y', position.y)\n",
       "              .attr('text-anchor', 'middle')\n",
       "              .attr('font-family', 'sans-serif')\n",
       "              .attr('font-size', '14px')\n",
       "              .style('font-weight', 'bold')\n",
       "              .attr('fill', 'red')\n",
       "              .attr('dy', '-1em')\n",
       "              .attr('transform', function(d) {\n",
       "                degrees = (angle + Math.PI) / (2 * Math.PI) * 360;\n",
       "                return 'rotate(' + (degrees - 90) + ' ' + (position.x) + ',' + (position.y) + ')';\n",
       "              })\n",
       "              .text(d.name);\n",
       "        })\n",
       "        .on('mouseout', function(d) {\n",
       "          d3.select('#lineFeatureHighlight').remove();\n",
       "        });\n",
       "    } else {\n",
       "\n",
       "    ///////////////////////////////////////\n",
       "    // Placement rules for linear sequences\n",
       "    ///////////////////////////////////////\n",
       "    var backboneHeight = 25;  // TODO: replace with unified thickness param\n",
       "    var xMin = config.padding;\n",
       "    var xMax = config.width - config.padding;\n",
       "\n",
       "    var backboneDim = {\n",
       "      'left': xMin,\n",
       "      'right': xMax,\n",
       "      'top': center.y - backboneHeight / 2,\n",
       "      'bottom': center.y + backboneHeight / 2\n",
       "    };\n",
       "\n",
       "    // Offset of 1 pixel makes the display look nicer\n",
       "    var xScale = d3.scale.linear()\n",
       "        .domain([0, sequence_json.sequence.length])\n",
       "        .range([xMin + 1, xMax + 1]);\n",
       "\n",
       "    ////////////////////\n",
       "    // Draw the backbone\n",
       "    ////////////////////\n",
       "    // TODO: similarities with circular - same fill, stroke, etc. Could simplify\n",
       "    //       with a backboneWidth attribute that's in pixels\n",
       "    svg.append('rect')\n",
       "        .attr('class', 'backbone')\n",
       "        .attr('x', backboneDim.left)\n",
       "        .attr('y', backboneDim.top)\n",
       "        .attr('width', backboneDim.right - backboneDim.left)\n",
       "        .attr('height', backboneDim.bottom - backboneDim.top)\n",
       "        .style('fill', config.backboneColor)\n",
       "        .style('stroke', d3.rgb(config.backboneColor).darker(0.5))\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    /////////////////////////////////\n",
       "    // Add plasmid name, number of bp\n",
       "    /////////////////////////////////\n",
       "    // TODO: similarities with circular - identical, just different placement\n",
       "    // FIXME: make the placement of text have no magic numbers\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.name)\n",
       "        .attr('x', center.x)\n",
       "        .attr('y', backboneDim.bottom +  50)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', config.titleFontSize);\n",
       "\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.sequence.length + ' bp')\n",
       "        .attr('x', center.x)\n",
       "        .attr('y', backboneDim.bottom + 80)\n",
       "        .style('text-anchor', 'middle');\n",
       "\n",
       "    /////////////\n",
       "    // Tick Marks\n",
       "    /////////////\n",
       "    // TODO: similarities with circular - almost everything\n",
       "    // Generate ticks data\n",
       "    var tickData = [];\n",
       "    for (var i=0; i < config.ticks; i++) {\n",
       "      tickData.push(i * Math.floor(sequence_json.sequence.length / config.ticks));\n",
       "    }\n",
       "\n",
       "    var tickGroup = svg.append('g')\n",
       "        .attr('id', 'tick-group');\n",
       "\n",
       "    var ticks = tickGroup.selectAll('.tick')\n",
       "        .data(tickData)\n",
       "      .enter().append('line')\n",
       "        .attr('x1', function(d) { return xScale(d); })\n",
       "        .attr('x2', function(d) { return xScale(d); })\n",
       "        .attr('y1', backboneDim.bottom)\n",
       "        .attr('y2', backboneDim.bottom + config.tickLength)\n",
       "        .attr('stroke', function(d, i) {\n",
       "          if (i % config.majorTickInterval === 0) {\n",
       "            tickColor = config.tickColorMajor;\n",
       "          } else {\n",
       "            tickColor = config.tickColorMinor;\n",
       "          }\n",
       "          return tickColor;\n",
       "        })\n",
       "        .attr('stroke-width', 2)\n",
       "        .attr('fill', 'none');\n",
       "\n",
       "    var tickLabels = tickGroup.selectAll('.tick-label')\n",
       "        .data(tickData)\n",
       "      .enter().append('text')\n",
       "        .text(function(d) {\n",
       "            if (d % config.majorTickInterval === 0) {\n",
       "              return d;\n",
       "            }\n",
       "        })\n",
       "        .attr('x', function(d) { return xScale(d); })\n",
       "        .attr('y', function(d) { return backboneDim.bottom + config.tickLength + 8; })\n",
       "        .attr('dy', '0.4em')\n",
       "        .attr('fill', config.tickColorMinor)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', '12px');\n",
       "\n",
       "    ////////////////////////////////\n",
       "    // Add features and their labels\n",
       "    ////////////////////////////////\n",
       "\n",
       "    // Separate out arc features and line features\n",
       "    var blockData = [];\n",
       "    var lineData = [];\n",
       "    // FIXME: this process is broken and should be determined by fit, not n\n",
       "    // arbitrary scaling factor\n",
       "    for (var i = 0; i < sequence_json.features.length; i++) {\n",
       "      feature = sequence_json.features[i];\n",
       "      if (xScale(feature.stop - feature.start) > 10 * config.minFeatureSize) {\n",
       "        blockData.push(feature);\n",
       "      } else {\n",
       "        lineData.push(feature);\n",
       "      }\n",
       "    }\n",
       "\n",
       "    var featuresGroup = svg.append('g')\n",
       "        .attr('id', 'features');\n",
       "    var blockGroup = featuresGroup.append('g')\n",
       "        .attr('id', 'block-features');\n",
       "\n",
       "    var blockFeatures = blockGroup.selectAll('.feature')\n",
       "        .data(blockData)\n",
       "      .enter().append('g')\n",
       "        .attr('class', 'feature')\n",
       "      .append('rect')\n",
       "        .attr('class', 'feature-block')\n",
       "        .attr('x', function(d) { return xScale(d.start); })\n",
       "        .attr('y', function(d) { return backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2; })\n",
       "        .attr('width', function(d) { return xScale(d.stop) - xScale(d.start); })\n",
       "        .attr('height', function(d) { return config.featureScale * backboneHeight; })\n",
       "        .attr('id', function(d, i) { return 'block' + i; })\n",
       "        .style('fill', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            return d.color;\n",
       "          } else {\n",
       "            return 'rgb(192, 192, 256)';\n",
       "          }\n",
       "        })\n",
       "        .style('fill-opacity', config.featureOpacity)\n",
       "        .style('stroke', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            color = d.color;\n",
       "          } else {\n",
       "            color = 'rgb(192, 192, 256)';\n",
       "          }\n",
       "          return d3.rgb(color).darker(0.5);\n",
       "        })\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    // Gaussian blur filter for making text shadow\n",
       "    var filter = svg.append('defs')\n",
       "      .append('filter')\n",
       "        .attr('id', 'text-shadow')\n",
       "        .attr('height', '130%')\n",
       "      .append('feGaussianBlur')\n",
       "        .attr('in', 'SourceAlpha')\n",
       "        .attr('stdDeviation', 2)\n",
       "        .attr('result', 'blur');\n",
       "\n",
       "    var shadows = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'shadow')\n",
       "        .attr('id', function(d, i) {\n",
       "          return 'shadow' + i;\n",
       "        })\n",
       "        .attr('x', function(d) { return xScale((d.start + d.stop) / 2); })\n",
       "        .attr('y', function(d) { return (backboneDim.top + backboneDim.bottom) / 2 + 3; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '10px')\n",
       "        .style('filter', 'url(#text-shadow)')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    var blockText = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-text')\n",
       "        .attr('fill', 'white')\n",
       "        .attr('id', function(d, i) { return 'label' + i; })\n",
       "        .attr('x', function(d) { return xScale((d.start + d.stop) / 2); })\n",
       "        .attr('y', function(d) { return (backboneDim.top + backboneDim.bottom) / 2; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '10px')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    // If text is too big, don't offset and set anchor to left\n",
       "    // FIXME: this should really be done with more d3-style methods\n",
       "    feature_selection = document.getElementsByClassName('feature');\n",
       "    for (var i = 0; i < feature_selection.length; i++) {\n",
       "      blocktemp = feature_selection[i];\n",
       "      text = blocktemp.getElementsByClassName('block-text')[0];\n",
       "      shadow = blocktemp.getElementsByClassName('shadow')[0];\n",
       "\n",
       "      block = blocktemp.getElementsByClassName('feature-block');\n",
       "      blockLength = block[0].width.baseVal.value\n",
       "\n",
       "      nameLength = text.innerHTML.length;\n",
       "      while (1.2 * text.getComputedTextLength() > blockLength) {\n",
       "         nameLength -= 1;\n",
       "         // Trim until it isn't\n",
       "         text.innerHTML = text.innerHTML.slice(0, nameLength);\n",
       "\n",
       "         shadow.innerHTML = shadow.innerHTML.slice(0, nameLength);\n",
       "       }\n",
       "    }\n",
       "\n",
       "    // Add line features\n",
       "    var lineGroup = featuresGroup.append('g')\n",
       "        .attr('class', 'lineFeatureGroup');\n",
       "\n",
       "    var lineFeatures = lineGroup.selectAll('.feature')\n",
       "        .data(lineData)\n",
       "      .enter().append('g')\n",
       "      .append('line')\n",
       "        .attr('class', 'feature')\n",
       "        .attr('class', 'line-feature')\n",
       "        .attr('x1', function(d) { return xScale(d.start); })\n",
       "        .attr('x2', function(d) { return xScale(d.start); })\n",
       "        .attr('y1', function(d) { return backboneDim.bottom + (config.featureScale * backboneHeight - backboneHeight) / 2 + config.tickLength; })\n",
       "        .attr('y2', function(d) { return backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2 - config.tickLength; })\n",
       "        .attr('stroke', 'rgb(96, 96, 96)')\n",
       "        .attr('stroke-width', config.lineFeatureWidth)\n",
       "        .attr('fill', 'none')\n",
       "        .on('mouseover', function(d) {\n",
       "          lineGroup.append('text')\n",
       "              .attr('id', 'lineFeatureHighlight')\n",
       "              .attr('x', xScale(d.start))\n",
       "              .attr('y', backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2 - 3)\n",
       "              .attr('text-anchor', 'middle')\n",
       "              .attr('font-family', 'sans-serif')\n",
       "              .attr('font-size', '14px')\n",
       "              .style('font-weight', 'bold')\n",
       "              .attr('fill', 'red')\n",
       "              .attr('dy', '-1em')\n",
       "              .text(d.name);\n",
       "        })\n",
       "        .on('mouseout', function(d) {\n",
       "          d3.select('#lineFeatureHighlight').remove();\n",
       "        });\n",
       "\n",
       "  }\n",
       "}\n",
       "\n",
       "            d3sequence(window.data, 'sequence_e0d77bc8-c47f-11e5-9eaa-02591d3bb8ad')\n",
       "        });\n",
       "        </script>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mch_feature = backbone_source.select_features('mCherry')[0]\n",
    "backbone = backbone_source.excise(mch_feature)\n",
    "backbone.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id=sequence_e1200884-c47f-11e5-9eaa-02591d3bb8ad></div>\n",
       "        <script>\n",
       "        require(['//d3js.org/d3.v3.min.js'], function(lib) {\n",
       "            window.data = {\"features\": [{\"start\": 0, \"stop\": 1605, \"name\": \"YJL014W\", \"type\": \"CDS\"}, {\"color\": [\"#b7e6d7\"], \"start\": 6051, \"stop\": 6075, \"name\": \"PP2\", \"type\": \"primer_bind\"}, {\"color\": [\"#c7b0e3\"], \"start\": 6075, \"stop\": 6730, \"name\": \"Actual GPD Promoter\", \"type\": \"promoter\"}, {\"color\": [\"#f8d3a9\"], \"start\": 1605, \"stop\": 1627, \"name\": \"TP\", \"type\": \"misc_feature\"}, {\"color\": [\"#f58a5e\"], \"start\": 1627, \"stop\": 1867, \"name\": \"CYC1_terminator\", \"type\": \"terminator\"}, {\"color\": [\"#b1ff67\"], \"start\": 1636, \"stop\": 1655, \"name\": \"pYESTrp_rev_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#f8d3a9\"], \"start\": 1636, \"stop\": 1655, \"name\": \"CYC1_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#9eafd2\"], \"start\": 1867, \"stop\": 1889, \"name\": \"TS\", \"type\": \"primer_bind\"}, {\"color\": [\"#ffef86\"], \"start\": 1889, \"stop\": 2389, \"name\": \"URA3 3' UTR\", \"type\": \"3'UTR\"}, {\"color\": [\"#b7e6d7\"], \"start\": 2389, \"stop\": 2397, \"name\": \"PmeI(1)\", \"type\": \"misc_feature\"}, {\"color\": [\"#d6b295\"], \"start\": 2401, \"stop\": 2418, \"name\": \"M13 Reverse primer\", \"type\": \"promoter\"}, {\"color\": [\"#c6c9d1\"], \"start\": 2414, \"stop\": 2437, \"name\": \"M13_pUC_rev_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#f8d3a9\"], \"start\": 2456, \"stop\": 2457, \"name\": \"mutant?\", \"type\": \"misc_feature\"}, {\"color\": [\"#75c6a9\"], \"start\": 2788, \"stop\": 3408, \"name\": \"pBR322_origin\", \"type\": \"rep_origin\"}, {\"color\": [\"#85dae9\"], \"start\": 2816, \"stop\": 2817, \"name\": \"mut?\", \"type\": \"misc_feature\"}, {\"color\": [\"#75c6a9\"], \"start\": 3562, \"stop\": 4423, \"name\": \"Ampicillin\", \"type\": \"CDS\"}, {\"color\": [\"#f8d3a9\"], \"start\": 4464, \"stop\": 4493, \"name\": \"AmpR_promoter\", \"type\": \"promoter\"}, {\"color\": [\"#c7b0e3\"], \"start\": 4651, \"stop\": 4674, \"name\": \"pGEX_3_primer\", \"type\": \"misc_feature\"}, {\"color\": [\"#85dae9\"], \"start\": 4809, \"stop\": 4817, \"name\": \"PmeI\", \"type\": \"misc_feature\"}, {\"color\": [\"#b1ff67\"], \"start\": 4817, \"stop\": 5065, \"name\": \"URA3 promoter (kind of)\", \"type\": \"promoter\"}, {\"color\": [\"#b4abac\"], \"start\": 5065, \"stop\": 5770, \"name\": \"URA3\", \"type\": \"gene\"}, {\"color\": [\"#85dae9\"], \"start\": 5765, \"stop\": 5766, \"name\": \"mutated to A in ura3-1 strains\", \"type\": \"misc_difference\"}, {\"color\": [\"#faac61\"], \"start\": 5869, \"stop\": 6051, \"name\": \"ADH1 Terminator\", \"type\": \"terminator\"}], \"bottom\": \"TTTGTAGATCCACTAGTTCTAGAATCCGATCCGTCGAAACTAAGTTCTGGTGTTTTAAAACTAAAAAAAAGACTAACTATAAAAGTAGAATTTAAGAAGTTTAAGAAATAGATTTACAGAATTACAATCAATACCTACCGTCTTTATATACTTATTAGTCAAGTAGGGGAATAATTTCAGGGAACTGGTTTCAACCTTTTTTTTCAGCTTTTTCCAAATCAGAGAGAGCAGAAGGTAATAGAAGGTGTAAGAAAATGAGATAGATACATGCGTGGGTCAATTGCCTTGTGTCATCATTTACTCCAGGCAGGTTGCATCACTCCATTGAGGTTGTGCCCGTTTTTTGCCTGTTTGTGCCCCTGTTCTCTGTAGTTGCGCTAAGAGAATGGACCTATGAACTGATGGTTGGTGAAGAAAACAATATTTTGGTGCTGGGATTCTTTTTTTTTCTGGATGCCAGCTTAAAAAGCGGGCTCCATTATATTTAGTGGATGCCAGGAATAAACTGTTCACCCAGACACCTACGATGTTATATATTCTGTGTAACCCGCCCCCTATTTTGGGCATGTACGGGTTACAGCAGAATTAAAAGGCTAATTTTTTGACTAAATAAAGTTAGGAAAATCACTACTATTAATTATTTACGTATTCTTTGAAATGGCGAGTATTGATAATGATAAACTGCCGCTTTTCGAATCTGGTTAAGGGAGGTGTGGTCAATAAGAGCGACCTCATGCTATACCTGAGAAAGCAACCTGACCTACAGGAAAGAGTTACTCAAGAATAAGAATTTTCGTTTTAAAACCTAAGAGTCACTTTAAAATTTGTATACACTTATTTTTTTTATAACTTATTTAATAATAAAAATCATAAATCATAAGAAATTCGCTTAGTTTTGCTGGCCGCATCTTCTCAAATATGCTTCCCAGCCTGCTTTTCTGTAACGTTCACCCTCTACCTTAGCATCCCTTCCCTTTGCAAATAGTCCTCTTCCAACAATAATAATGTCAGATCCTGTAGAGACCACATCATCCACGGTTCTATACTGTTGACCCAATGCGTCTCCCTTGTCATCTAAACCCACACCGGGTGTCATAATCAACCAATCGTAACCTTCATCTCTTCCACCCATGTCTCTTTGAGCAATAAAGCCGATAACAAAATCTTTGTCGCTCTTCGCAATGTCAACAGTACCCTTAGTATATTCTCCAGTAGCTAGGGAGCCCTTGCATGACAATTCTGCTAACATCAAAAGGCCTCTAGGTTCCTTTGTTACTTCTTCCGCCGCCTGCTTCAAACCGCTAACAATACCTGGGCCCACCACACCGTGTGCATTCGTAATGTCTGCCCATTCTGCTATTCTGTATACACCCGCAGAGTACTGCAATTTGACTGTATTACCAATGTCAGCAAATTTTCTGTCTTCGAAGAGTAAAAAATTGTACTTGGCGGATAATGCCTTTAGCGGCTTAACTGTGCCCTCCATGGAAAAATCAGTCAAGATATCCACATGTGTTTTTAGTAAACAAATTTTGGGACCTAATGCTTCAACTAACTCCAGTAATTCCTTGGTGGTACGAACATCCAATGAAGCACACAAGTTTGTTTGCTTTTCGTGCATGATATTAAATAGCTTGGCAGCAACAGGACTAGGATGAGTAGCAGCACGTTCCTTATATGTAGCTTTCGACATGATTTATCTTCGTTTCCTGCAGGTTTTTGTTCTGTGCAGTTGGGTTAAGAATACTGGGCAATTTCATGTTTCTTCAACACCACATATGCGTATATATACCAATCTAAGTCTGTGCTCCTTCCTTCGTTCTTCCTTCTGCTCGGAGATTACCGAATCAAAAAAATTTCAAAGAAACCGGAATCAAAAAAAAGAACAAAAAAAAAAAAGATGAATTGAAAAGCTTTATGGACCCTGAAACCACAGCCACAGTTTAAACTATGGTGCACTCTCAGTACAATCTGCTCTGATGCCGCATAGTTAAGCCAGCCCCGACACCCGCCAACACCCGCTGACGCGCCCTGACGGGCTTGTCTGCTCCCGGCATCCGCTTACAGACAAGCTGTGACCGTCTCCGGGAGCTGCATGTGTCAGAGGTTTTCACCGTCATCACCGAAACGCGCGAGACGAAAGGGCCTCGTGATACGCCTATTTTTATAGGTTAATGTCATGATAATAATGGTTTCTTAGACGTCAGGTGGCACTTTTCGGGGAAATGTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATTCAAATATGTATCCGCTCATGAGACAATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAGTATGAGTATTCAACATTTCCGTGTCGCCCTTATTCCCTTTTTTGCGGCATTTTGCCTTCCTGTTTTTGCTCACCCAGAAACGCTGGTGAAAGTAAAAGATGCTGAAGATCAGTTGGGTGCACGAGTGGGTTACATCGAACTGGATCTCAACAGCGGTAAGATCCTTGAGAGTTTTCGCCCCGAAGAACGTTTTCCAATGATGAGCACTTTTAAAGTTCTGCTATGTGGCGCGGTATTATCCCGTATTGACGCCGGGCAAGAGCAACTCGGTCGCCGCATACACTATTCTCAGAATGACTTGGTTGAGTACTCACCAGTCACAGAAAAGCATCTTACGGATGGCATGACAGTAAGAGAATTATGCAGTGCTGCCATAACCATGAGTGATAACACTGCGGCCAACTTACTTCTGACAACGATCGGAGGACCGAAGGAGCTAACCGCTTTTTTTCACAACATGGGGGATCATGTAACTCGCCTTGATCGTTGGGAACCGGAGCTGAATGAAGCCATACCAAACGACGAGCGTGACACCACGATGCCTGTAGCAATGGCAACAACGTTGCGCAAACTATTAACTGGCGAACTACTTACTCTAGCTTCCCGGCAACAATTAATAGACTGGATGGAGGCGGATAAAGTTGCAGGACCACTTCTGCGCTCGGCCCTTCCGGCTGGCTGGTTTATTGCTGATAAATCTGGAGCCGGTGAGCGTGGGTCTCGCGGTATCATTGCAGCACTGGGGCCAGATGGTAAGCCCTCCCGTATCGTAGTTATCTACACGACGGGCAGTCAGGCAACTATGGATGAACGAAATAGACAGATCGCTGAGATAGGTGCCTCACTGATTAAGCATTGGTAACTGTCAGACCAAGTTTACTCATATATACTTTAGATTGATTTAAAACTTCATTTTTAATTTAAAAGGATCTAGGTGAAGATCCTTTTTGATAATCTCATGACCAAAATCCCTTAACGTGAGTTTTCGTTCCACTGAGCGTCAGACCCCGTAGAAAAGATCAAAGGATCTTCTTGAGATCCTTTTTTTCTGCGCGTAATCTGCTGCTTGCAAACAAAAAAACCACCGCTACCAGCGGTGGTTTGTTTGCCGGATCAAGAGCTACCAACTCTTTTTCCGAAGGTAACTGGCTTCAGCAGAGCGCAGATACCAAATACTGTCCTTCTAGTGTAGCCGTAGTTAGGCCACCACTTCAAGAACTCTGTAGCACCGCCTACATACCTCGCTCTGCTAATCCTGTTACCAGTGGCTGCTGCCAGTGGCGATAAGTCGTGTCTTACCGGGTTGGACTCAAGACGATAGTTACCGGATAAGGCGCAGCGGTCGGGCTGAACGGGGGGTTCGTGCACACAGCCCAGCTTGGAGCGAACGACCTACACCGAACTGAGATACCTACAGCGTGAGCATTGAGAAAGCGCCACGCTTCCCGAAGGGAGAAAGGCGGACAGGTATCCGGTAAGCGGCAGGGTCGGAACAGGAGAGCGCACGAGGGAGCTTCCAGGGGGGAACGCCTGGTATCTTTATAGTCCTGTCGGGTTTCGCCACCTCTGACTTGAGCGTCGATTTTTGTGATGCTCGTCAGGGGGGCGGAGCCTATGGAAAAACGCCAGCAACGCGGCCTTTTTACGGTTCCTGGCCTTTTGCTGGCCTTTTGCTCACATGTTCTTTCCTGCGTTATCCCCTGATTCTGTGGATAACCGTATTACCGCCTTTGAGTGAGCTGATACCGCTCGCCGCAGCCGAACGACCGAGCGCAGCGAGTCAGTGAGCGAGGAAGCGGAAGAGCGCCCAATACGCAAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTGAGCGCAACGCAATTAATGTGAGTTACCTCACTCATTAGGCACCCCAGGCTTTACACTTTATGCTTCCGGCTCGTATGTTGTGTGGAATTGTGAGCGGATAACAATTTCACACAGGAAACAGCTATGACCATGGTTTAAACCGGCCCAAGCCTTGTCCCAAGGCAGCGTTTTGTTCTTGGAAACGCTGCCCTACACGTTCGCTATGCTTCAAGAACTTTTCTGAGCACTTCATGATGCATGTTTGTTCCTTATTGGTTAGCTTTGATGTTGTGAAGTCATTGACACAGTCTGTGAAACATCTTTCTACCAGATTAGAGTACAAACGCATGAAATCCTTCATTTGCTTTTGTTCCACTACTTTTTGGAACTCTTGTTGTTCTTTGGAGTTCAATGCGTCCATCTTTACAGTCCTGTCTTATTGTTCTTGATTTGTGCCCCGTAAAATACTGTTACTTGGTTCTGGCGAGGTATTGGATAGTTCCTTTTTATAAAGGCCATGAAGCTTTTTCTTTCCAATTTTTTTTTTTTCGTCATTATAGAAATCATTACGACCGAGATTCCCGGGTAATAACTGATATAATTAAATTGAAGCTCTAATTTGTGAGTTTAGTATACATGCATTTACTTATAATACAGTTTTCGGGGCGGGGTTATTACGACATCAAATTAAAGCCTTCGAGCGTCCCAAAACCTTCTCAAGCAAGGTTTTCAGTATAATGTTACATGCGTACACGCGTCTGTACAGAAAAAAAAGAAAAATTTGAAATATAAATAACGTTCTTAATACTAACATAACTATAAAAAAATAAATAGGGACCTAGACTTCAGGTTGTCTAACTCCTTCCTTTTCGGTTAGAGCGGATGTGGGGGGAGGGCGTGAATGTAAGCGTGACATAACTAATTGACTCGAGGTCGACGGTATCATTATTCTTGTTTTCTTACACCACTAACAATATCATCAACCCTTAATAGTAAACATGCACTTTCAATAGCGGTTTTGACACTTTGCTGCTTGATGACTTCAGGTTCCCAAATACCGTAACTAACCATATCTACAATTTTCCCCTTGTCACCATCAATACCTGTTGTAAAATTTCCCTGTGCGTGTTTGGCTCTTAGTTGGGATAGGAGTCTAATAGGATCACCACCTGCATTTTGTATCAAAGTACGTGGAATACACTCCATAGCATCAGCTACAGCTTGGTATGGCCATTGCTGGATCCCTTCTAATTGCTTTGCCTTTTCTGCTAATTTGACAGAAACTGCCATCTCGGTGGCACCACCACCTGGGGACAAGGAAGGACTCAACATCACATTACGAGCAACAGCCATAGCATCTTGTAGGTTACGATCTATTTCGTTTAATATATCTTTGGAACCACCACGTAACATTATGGTACAAGCTTTTGGTTCCTTGCAGTTATCCAAAAATGAGAAATATTCGTCCCCAATCATTTCTACTTTGAATAATCCACAGTTTGTACCGACATCACTTTCCTTCAAGTCCTCTACACGATTGACAATGGTTGCACCAGTTACACGAGCAATTCTATTATTGTCTGATTTCTTCACTCTTCTTAAAACACTACATCCACCCTTTAGTAAATAATGCTGTGCTAGATCTGAAACACCTTTTTCAGTGATGACTAGTGTTGGCCTAACGGCTAAAATTTGTTCACACATTAGCTGAACTTGTTCTTCCTCGATTTGTAAAATCCTATTCCAATCTTCCTCCTTTTCAATTTCTATGTTGGTTTGAGATTCACCCTTCTTATATTCTAGTGGACAATCCAAAAGAACAACACGTGGATTCTCTATGTGACGGGACATCTTTGGATGAACAACATCCTTGTTCAATAAGACACCCTTTAAGACGCGAGAATCTAGAACATCCCCCCCAGGAATCTTCTCCACACGGACATATCTTTTGATATCGATTTCAAAATTAGGTTCACCTTCCACAGTTTGTCCCAGGTCTTTACGGACAGTCTTAACAGCGTCTAGAGCTAATTCACACATTTTCTCTGACCAATGTATGACATATTTAGTACCGATAGAGGCTTGAATCAATTTTTTCATAGCGGCATCATTTTCCACATCGACAGGTTTACTTACTTGTTTGATAACTTCTAGTGCATCAGTCAGTGCTTTCTTCAAGGCTTGGATAATAATAACAGGGTGAATATTTTTTTCAATCAAATAAGGTGCACACTGAGCTAAAATCTCACCAGCTAGAATAATAACTGTTGTAGTCCCGTCACCAACCTCTTCGTCTTGAGTACGGGATAACTCCAACATTGATTTAGCTGCTGGATGTGCAACATCAATTTCCCTTAAAATAGCGTGGCCATCGTTAGTCAACACAAGACCGCCCATAGGATCCAATAACATCTTTAACATAGCTTTTGGACCTAAGCAAGTTCGAATGACATCGGCAACTGCCTTTGCAGCAGTGATATTGGAGATTTGGGCCTGACGACCGGTAGTTCTCTCTTGAGAAGCATTCATAAATACCACTGGAGCTTGCAT\", \"material\": \"DNA\", \"sequence\": \"ATGCAAGCTCCAGTGGTATTTATGAATGCTTCTCAAGAGAGAACTACCGGTCGTCAGGCCCAAATCTCCAATATCACTGCTGCAAAGGCAGTTGCCGATGTCATTCGAACTTGCTTAGGTCCAAAAGCTATGTTAAAGATGTTATTGGATCCTATGGGCGGTCTTGTGTTGACTAACGATGGCCACGCTATTTTAAGGGAAATTGATGTTGCACATCCAGCAGCTAAATCAATGTTGGAGTTATCCCGTACTCAAGACGAAGAGGTTGGTGACGGGACTACAACAGTTATTATTCTAGCTGGTGAGATTTTAGCTCAGTGTGCACCTTATTTGATTGAAAAAAATATTCACCCTGTTATTATTATCCAAGCCTTGAAGAAAGCACTGACTGATGCACTAGAAGTTATCAAACAAGTAAGTAAACCTGTCGATGTGGAAAATGATGCCGCTATGAAAAAATTGATTCAAGCCTCTATCGGTACTAAATATGTCATACATTGGTCAGAGAAAATGTGTGAATTAGCTCTAGACGCTGTTAAGACTGTCCGTAAAGACCTGGGACAAACTGTGGAAGGTGAACCTAATTTTGAAATCGATATCAAAAGATATGTCCGTGTGGAGAAGATTCCTGGGGGGGATGTTCTAGATTCTCGCGTCTTAAAGGGTGTCTTATTGAACAAGGATGTTGTTCATCCAAAGATGTCCCGTCACATAGAGAATCCACGTGTTGTTCTTTTGGATTGTCCACTAGAATATAAGAAGGGTGAATCTCAAACCAACATAGAAATTGAAAAGGAGGAAGATTGGAATAGGATTTTACAAATCGAGGAAGAACAAGTTCAGCTAATGTGTGAACAAATTTTAGCCGTTAGGCCAACACTAGTCATCACTGAAAAAGGTGTTTCAGATCTAGCACAGCATTATTTACTAAAGGGTGGATGTAGTGTTTTAAGAAGAGTGAAGAAATCAGACAATAATAGAATTGCTCGTGTAACTGGTGCAACCATTGTCAATCGTGTAGAGGACTTGAAGGAAAGTGATGTCGGTACAAACTGTGGATTATTCAAAGTAGAAATGATTGGGGACGAATATTTCTCATTTTTGGATAACTGCAAGGAACCAAAAGCTTGTACCATAATGTTACGTGGTGGTTCCAAAGATATATTAAACGAAATAGATCGTAACCTACAAGATGCTATGGCTGTTGCTCGTAATGTGATGTTGAGTCCTTCCTTGTCCCCAGGTGGTGGTGCCACCGAGATGGCAGTTTCTGTCAAATTAGCAGAAAAGGCAAAGCAATTAGAAGGGATCCAGCAATGGCCATACCAAGCTGTAGCTGATGCTATGGAGTGTATTCCACGTACTTTGATACAAAATGCAGGTGGTGATCCTATTAGACTCCTATCCCAACTAAGAGCCAAACACGCACAGGGAAATTTTACAACAGGTATTGATGGTGACAAGGGGAAAATTGTAGATATGGTTAGTTACGGTATTTGGGAACCTGAAGTCATCAAGCAGCAAAGTGTCAAAACCGCTATTGAAAGTGCATGTTTACTATTAAGGGTTGATGATATTGTTAGTGGTGTAAGAAAACAAGAATAATGATACCGTCGACCTCGAGTCAATTAGTTATGTCACGCTTACATTCACGCCCTCCCCCCACATCCGCTCTAACCGAAAAGGAAGGAGTTAGACAACCTGAAGTCTAGGTCCCTATTTATTTTTTTATAGTTATGTTAGTATTAAGAACGTTATTTATATTTCAAATTTTTCTTTTTTTTCTGTACAGACGCGTGTACGCATGTAACATTATACTGAAAACCTTGCTTGAGAAGGTTTTGGGACGCTCGAAGGCTTTAATTTGATGTCGTAATAACCCCGCCCCGAAAACTGTATTATAAGTAAATGCATGTATACTAAACTCACAAATTAGAGCTTCAATTTAATTATATCAGTTATTACCCGGGAATCTCGGTCGTAATGATTTCTATAATGACGAAAAAAAAAAAATTGGAAAGAAAAAGCTTCATGGCCTTTATAAAAAGGAACTATCCAATACCTCGCCAGAACCAAGTAACAGTATTTTACGGGGCACAAATCAAGAACAATAAGACAGGACTGTAAAGATGGACGCATTGAACTCCAAAGAACAACAAGAGTTCCAAAAAGTAGTGGAACAAAAGCAAATGAAGGATTTCATGCGTTTGTACTCTAATCTGGTAGAAAGATGTTTCACAGACTGTGTCAATGACTTCACAACATCAAAGCTAACCAATAAGGAACAAACATGCATCATGAAGTGCTCAGAAAAGTTCTTGAAGCATAGCGAACGTGTAGGGCAGCGTTTCCAAGAACAAAACGCTGCCTTGGGACAAGGCTTGGGCCGGTTTAAACCATGGTCATAGCTGTTTCCTGTGTGAAATTGTTATCCGCTCACAATTCCACACAACATACGAGCCGGAAGCATAAAGTGTAAAGCCTGGGGTGCCTAATGAGTGAGGTAACTCACATTAATTGCGTTGCGCTCACTGCCCGCTTTCCAGTCGGGAAACCTGTCGTGCCAGCTGCATTAATGAATCGGCCAACGCGCGGGGAGAGGCGGTTTGCGTATTGGGCGCTCTTCCGCTTCCTCGCTCACTGACTCGCTGCGCTCGGTCGTTCGGCTGCGGCGAGCGGTATCAGCTCACTCAAAGGCGGTAATACGGTTATCCACAGAATCAGGGGATAACGCAGGAAAGAACATGTGAGCAAAAGGCCAGCAAAAGGCCAGGAACCGTAAAAAGGCCGCGTTGCTGGCGTTTTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAAAATCGACGCTCAAGTCAGAGGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTCCCCCCTGGAAGCTCCCTCGTGCGCTCTCCTGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGGAAGCGTGGCGCTTTCTCAATGCTCACGCTGTAGGTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTGGGCTGTGTGCACGAACCCCCCGTTCAGCCCGACCGCTGCGCCTTATCCGGTAACTATCGTCTTGAGTCCAACCCGGTAAGACACGACTTATCGCCACTGGCAGCAGCCACTGGTAACAGGATTAGCAGAGCGAGGTATGTAGGCGGTGCTACAGAGTTCTTGAAGTGGTGGCCTAACTACGGCTACACTAGAAGGACAGTATTTGGTATCTGCGCTCTGCTGAAGCCAGTTACCTTCGGAAAAAGAGTTGGTAGCTCTTGATCCGGCAAACAAACCACCGCTGGTAGCGGTGGTTTTTTTGTTTGCAAGCAGCAGATTACGCGCAGAAAAAAAGGATCTCAAGAAGATCCTTTGATCTTTTCTACGGGGTCTGACGCTCAGTGGAACGAAAACTCACGTTAAGGGATTTTGGTCATGAGATTATCAAAAAGGATCTTCACCTAGATCCTTTTAAATTAAAAATGAAGTTTTAAATCAATCTAAAGTATATATGAGTAAACTTGGTCTGACAGTTACCAATGCTTAATCAGTGAGGCACCTATCTCAGCGATCTGTCTATTTCGTTCATCCATAGTTGCCTGACTGCCCGTCGTGTAGATAACTACGATACGGGAGGGCTTACCATCTGGCCCCAGTGCTGCAATGATACCGCGAGACCCACGCTCACCGGCTCCAGATTTATCAGCAATAAACCAGCCAGCCGGAAGGGCCGAGCGCAGAAGTGGTCCTGCAACTTTATCCGCCTCCATCCAGTCTATTAATTGTTGCCGGGAAGCTAGAGTAAGTAGTTCGCCAGTTAATAGTTTGCGCAACGTTGTTGCCATTGCTACAGGCATCGTGGTGTCACGCTCGTCGTTTGGTATGGCTTCATTCAGCTCCGGTTCCCAACGATCAAGGCGAGTTACATGATCCCCCATGTTGTGAAAAAAAGCGGTTAGCTCCTTCGGTCCTCCGATCGTTGTCAGAAGTAAGTTGGCCGCAGTGTTATCACTCATGGTTATGGCAGCACTGCATAATTCTCTTACTGTCATGCCATCCGTAAGATGCTTTTCTGTGACTGGTGAGTACTCAACCAAGTCATTCTGAGAATAGTGTATGCGGCGACCGAGTTGCTCTTGCCCGGCGTCAATACGGGATAATACCGCGCCACATAGCAGAACTTTAAAAGTGCTCATCATTGGAAAACGTTCTTCGGGGCGAAAACTCTCAAGGATCTTACCGCTGTTGAGATCCAGTTCGATGTAACCCACTCGTGCACCCAACTGATCTTCAGCATCTTTTACTTTCACCAGCGTTTCTGGGTGAGCAAAAACAGGAAGGCAAAATGCCGCAAAAAAGGGAATAAGGGCGACACGGAAATGTTGAATACTCATACTCTTCCTTTTTCAATATTATTGAAGCATTTATCAGGGTTATTGTCTCATGAGCGGATACATATTTGAATGTATTTAGAAAAATAAACAAATAGGGGTTCCGCGCACATTTCCCCGAAAAGTGCCACCTGACGTCTAAGAAACCATTATTATCATGACATTAACCTATAAAAATAGGCGTATCACGAGGCCCTTTCGTCTCGCGCGTTTCGGTGATGACGGTGAAAACCTCTGACACATGCAGCTCCCGGAGACGGTCACAGCTTGTCTGTAAGCGGATGCCGGGAGCAGACAAGCCCGTCAGGGCGCGTCAGCGGGTGTTGGCGGGTGTCGGGGCTGGCTTAACTATGCGGCATCAGAGCAGATTGTACTGAGAGTGCACCATAGTTTAAACTGTGGCTGTGGTTTCAGGGTCCATAAAGCTTTTCAATTCATCTTTTTTTTTTTTGTTCTTTTTTTTGATTCCGGTTTCTTTGAAATTTTTTTGATTCGGTAATCTCCGAGCAGAAGGAAGAACGAAGGAAGGAGCACAGACTTAGATTGGTATATATACGCATATGTGGTGTTGAAGAAACATGAAATTGCCCAGTATTCTTAACCCAACTGCACAGAACAAAAACCTGCAGGAAACGAAGATAAATCATGTCGAAAGCTACATATAAGGAACGTGCTGCTACTCATCCTAGTCCTGTTGCTGCCAAGCTATTTAATATCATGCACGAAAAGCAAACAAACTTGTGTGCTTCATTGGATGTTCGTACCACCAAGGAATTACTGGAGTTAGTTGAAGCATTAGGTCCCAAAATTTGTTTACTAAAAACACATGTGGATATCTTGACTGATTTTTCCATGGAGGGCACAGTTAAGCCGCTAAAGGCATTATCCGCCAAGTACAATTTTTTACTCTTCGAAGACAGAAAATTTGCTGACATTGGTAATACAGTCAAATTGCAGTACTCTGCGGGTGTATACAGAATAGCAGAATGGGCAGACATTACGAATGCACACGGTGTGGTGGGCCCAGGTATTGTTAGCGGTTTGAAGCAGGCGGCGGAAGAAGTAACAAAGGAACCTAGAGGCCTTTTGATGTTAGCAGAATTGTCATGCAAGGGCTCCCTAGCTACTGGAGAATATACTAAGGGTACTGTTGACATTGCGAAGAGCGACAAAGATTTTGTTATCGGCTTTATTGCTCAAAGAGACATGGGTGGAAGAGATGAAGGTTACGATTGGTTGATTATGACACCCGGTGTGGGTTTAGATGACAAGGGAGACGCATTGGGTCAACAGTATAGAACCGTGGATGATGTGGTCTCTACAGGATCTGACATTATTATTGTTGGAAGAGGACTATTTGCAAAGGGAAGGGATGCTAAGGTAGAGGGTGAACGTTACAGAAAAGCAGGCTGGGAAGCATATTTGAGAAGATGCGGCCAGCAAAACTAAGCGAATTTCTTATGATTTATGATTTTTATTATTAAATAAGTTATAAAAAAAATAAGTGTATACAAATTTTAAAGTGACTCTTAGGTTTTAAAACGAAAATTCTTATTCTTGAGTAACTCTTTCCTGTAGGTCAGGTTGCTTTCTCAGGTATAGCATGAGGTCGCTCTTATTGACCACACCTCCCTTAACCAGATTCGAAAAGCGGCAGTTTATCATTATCAATACTCGCCATTTCAAAGAATACGTAAATAATTAATAGTAGTGATTTTCCTAACTTTATTTAGTCAAAAAATTAGCCTTTTAATTCTGCTGTAACCCGTACATGCCCAAAATAGGGGGCGGGTTACACAGAATATATAACATCGTAGGTGTCTGGGTGAACAGTTTATTCCTGGCATCCACTAAATATAATGGAGCCCGCTTTTTAAGCTGGCATCCAGAAAAAAAAAGAATCCCAGCACCAAAATATTGTTTTCTTCACCAACCATCAGTTCATAGGTCCATTCTCTTAGCGCAACTACAGAGAACAGGGGCACAAACAGGCAAAAAACGGGCACAACCTCAATGGAGTGATGCAACCTGCCTGGAGTAAATGATGACACAAGGCAATTGACCCACGCATGTATCTATCTCATTTTCTTACACCTTCTATTACCTTCTGCTCTCTCTGATTTGGAAAAAGCTGAAAAAAAAGGTTGAAACCAGTTCCCTGAAATTATTCCCCTACTTGACTAATAAGTATATAAAGACGGTAGGTATTGATTGTAATTCTGTAAATCTATTTCTTAAACTTCTTAAATTCTACTTTTATAGTTAGTCTTTTTTTTAGTTTTAAAACACCAGAACTTAGTTTCGACGGATCGGATTCTAGAACTAGTGGATCTACAAA\", \"circular\": true, \"name\": \"\"};// Full JSON description of plasmid - in the future, this would be\n",
       "// downloaded dynamically\n",
       "function d3sequence(sequence_json, div_id) {\n",
       "  // TODO: Have methods for drawing circular vs. linear, keep core logic the same\n",
       "  // Option that might make it into a configurable version\n",
       "  // TODO: control more of these with CSS\n",
       "\n",
       "  var circularConfig = {\n",
       "    'width': 500,\n",
       "    'height': 500,\n",
       "    'padding': 75,\n",
       "    'featureScale': 1.2,\n",
       "    'featureOpacity': 0.5,\n",
       "    'backboneThickness': 0.2,\n",
       "    'backboneColor': '#eee',\n",
       "    'minFeatureSize': 5,\n",
       "    'ticks': 80,\n",
       "    'majorTickInterval': 5,\n",
       "    'tickColorMinor': '#94aab0',\n",
       "    'tickColorMajor': 'black',\n",
       "    'tickLength': 6,\n",
       "    'lineFeatureWidth': 4,\n",
       "    'titleFontSize': '25px'\n",
       "  };\n",
       "\n",
       "  var linearConfig = {\n",
       "    'width': 800,\n",
       "    'height': 150,\n",
       "    'padding': 35,\n",
       "    'featureScale': 1.2,\n",
       "    'featureOpacity': 0.5,\n",
       "    'backboneThickness': 0.2,\n",
       "    'backboneColor': '#eee',\n",
       "    'minFeatureSize': 5,\n",
       "    'ticks': 40,\n",
       "    'majorTickInterval': 5,\n",
       "    'tickColorMinor': '#94aab0',\n",
       "    'tickColorMajor': 'black',\n",
       "    'tickLength': 6,\n",
       "    'lineFeatureWidth': 4,\n",
       "    'titleFontSize': '25px'\n",
       "  };\n",
       "\n",
       "  if (!sequence_json.circular) {\n",
       "    config = linearConfig;\n",
       "  } else if (sequence_json.circular) {\n",
       "    config = circularConfig;\n",
       "  }\n",
       "\n",
       "  ///////////////////////\n",
       "  // Initialize svg space\n",
       "  ///////////////////////\n",
       "  var svg = d3.select('#' + div_id)\n",
       "    .append('svg')\n",
       "      .attr('width', config.width)\n",
       "      .attr('height', config.height)\n",
       "    .append('g');\n",
       "\n",
       "  var center = {'x': config.width / 2, 'y': config.height / 2};\n",
       "\n",
       "  if (sequence_json.circular) {\n",
       "\n",
       "    var outerRadius = Math.min(config.width, config.height) / 2 - config.padding;\n",
       "    var innerRadius = outerRadius - config.backboneThickness * outerRadius;\n",
       "    var centerRadius = d3.mean([outerRadius, innerRadius]);\n",
       "    var featureOffset = (config.featureScale * (outerRadius - innerRadius) - (outerRadius - innerRadius)) / 2;\n",
       "\n",
       "    /////////////////////////////////////////\n",
       "    // Placement rules for circular sequences\n",
       "    /////////////////////////////////////////\n",
       "    var centerTranslation = 'translate(' + center.x + ',' + center.y + ')';\n",
       "    svg.attr('transform', centerTranslation);\n",
       "\n",
       "    var baseToRadians = d3.scale.linear()\n",
       "        .domain([0, sequence_json.sequence.length])\n",
       "        .range([-(Math.PI / 2), 1.5 * Math.PI]);\n",
       "\n",
       "    // Placement around the plasmid is always prior to centering in the svg\n",
       "    function polarToCartesian(radius, angle) {\n",
       "      return {'x': radius * Math.cos(angle), 'y': radius * Math.sin(angle)};\n",
       "    }\n",
       "\n",
       "\n",
       "    ///////////////////////////////////////////\n",
       "    // Make an annulus for the plasmid backbone\n",
       "    ///////////////////////////////////////////\n",
       "\n",
       "    var annulus = d3.svg.arc()\n",
       "        .innerRadius(innerRadius)\n",
       "        .outerRadius(outerRadius)\n",
       "        .startAngle(0)\n",
       "        .endAngle(2 * Math.PI);\n",
       "\n",
       "    backbone = svg.append('path')\n",
       "        .attr('class', 'backbone')\n",
       "        .attr('d', annulus)\n",
       "        .style('fill-rule', 'evenodd')\n",
       "        .style('fill', config.backboneColor)\n",
       "        .style('stroke', d3.rgb(config.backboneColor).darker(0.5))\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    /////////////////////////////////\n",
       "    // Add plasmid name, number of bp\n",
       "    /////////////////////////////////\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.name)\n",
       "        .attr('x', 0)\n",
       "        .attr('y', 0)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', config.titleFontSize);\n",
       "\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.sequence.length + ' bp')\n",
       "        .attr('x', 0)\n",
       "        .attr('y', 40)\n",
       "        .style('text-anchor', 'middle');\n",
       "\n",
       "    /////////////\n",
       "    // Tick Marks\n",
       "    /////////////\n",
       "\n",
       "    // Generate ticks data\n",
       "    var tickData = [];\n",
       "    for (var i=0; i < config.ticks; i++) {\n",
       "      tickData.push(i * Math.floor(sequence_json.sequence.length / config.ticks));\n",
       "    }\n",
       "\n",
       "    // FIXME: why use paths to make Ticks when lines can be used?\n",
       "    var line = d3.svg.line()\n",
       "        .x(function(d) { return d.x; } )\n",
       "        .y(function(d) { return d.y; } );\n",
       "\n",
       "    tickGroup = svg.append('g')\n",
       "        .attr('id', 'tick-group');\n",
       "\n",
       "    ticks = tickGroup.selectAll('.tick')\n",
       "        .data(tickData)\n",
       "      .enter().append('path')\n",
       "        .attr('d', function(d, i) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          inner = polarToCartesian(outerRadius, tickAngle);\n",
       "          outer = polarToCartesian(outerRadius + config.tickLength, tickAngle);\n",
       "          return line([inner, outer]);\n",
       "        })\n",
       "        .attr('stroke', function(d, i) {\n",
       "          if (i === 0) {\n",
       "            tickColor = 'red';\n",
       "          } else if (i % config.majorTickInterval === 0) {\n",
       "            tickColor = config.tickColorMajor;\n",
       "          } else {\n",
       "            tickColor = config.tickColorMinor;\n",
       "          }\n",
       "          return tickColor;\n",
       "        })\n",
       "        .attr('stroke-width', 2)\n",
       "        .attr('fill', 'none');\n",
       "\n",
       "    tickLabels = tickGroup.selectAll('.tick-label')\n",
       "        .data(tickData)\n",
       "      .enter().append('text')\n",
       "        .text(function(d) {\n",
       "            if (d % config.majorTickInterval === 0) {\n",
       "              return d;\n",
       "            }\n",
       "        })\n",
       "        .attr('x', function(d) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          return polarToCartesian(outerRadius + 22, tickAngle).x;\n",
       "        })\n",
       "        .attr('y', function(d) {\n",
       "          tickAngle = baseToRadians(d);\n",
       "          return polarToCartesian(outerRadius + 22, tickAngle).y;\n",
       "        })\n",
       "        .attr('dy', '0.4em')\n",
       "        .attr('fill', config.tickColorMinor)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', '12px');\n",
       "\n",
       "    ////////////////////////////////\n",
       "    // Add features and their labels\n",
       "    ////////////////////////////////\n",
       "\n",
       "    // Separate out arc features and line features\n",
       "    var blockData = [];\n",
       "    var lineData = [];\n",
       "    for (var i = 0; i < sequence_json.features.length; i++) {\n",
       "      feature = sequence_json.features[i];\n",
       "      arcAngle = 2 * Math.PI / sequence_json.sequence.length * (feature.stop - feature.start);\n",
       "      arcLength = centerRadius * arcAngle;\n",
       "      if (arcLength > config.minFeatureSize) {\n",
       "        blockData.push(feature);\n",
       "      } else {\n",
       "        lineData.push(feature);\n",
       "      }\n",
       "    }\n",
       "\n",
       "    var featuresGroup = svg.append('g')\n",
       "        .attr('id', 'features');\n",
       "    var blockGroup = featuresGroup.append('g')\n",
       "        .attr('id', 'block-features');\n",
       "\n",
       "    var blockFeatures = blockGroup.selectAll('.feature')\n",
       "        .data(blockData)\n",
       "      .enter()\n",
       "        .append('g')\n",
       "        .attr('class', 'feature')\n",
       "        .append('path')\n",
       "        .attr('d', function(d, i) {\n",
       "          the_arc = d3.svg.arc()\n",
       "            // d3 svg arc origin is at the top\n",
       "            .startAngle(baseToRadians(d.start) + Math.PI / 2)\n",
       "            .endAngle(baseToRadians(d.stop) + Math.PI / 2)\n",
       "            .innerRadius(innerRadius - featureOffset)\n",
       "            .outerRadius(outerRadius + featureOffset)\n",
       "            (d, i);\n",
       "          return the_arc;\n",
       "        })\n",
       "        .attr('id', function(d, i) { return 'block' + i; })\n",
       "        .style('fill-rule', 'evenodd')\n",
       "        .style('fill', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            return d.color;\n",
       "          } else {\n",
       "            return 'rgb(192, 192, 256)';\n",
       "          }\n",
       "        })\n",
       "        .style('fill-opacity', config.featureOpacity)\n",
       "        .style('stroke', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            color = d.color;\n",
       "          } else {\n",
       "            color = 'rgb(192, 192, 256)';\n",
       "          }\n",
       "          return d3.rgb(color).darker(0.5);\n",
       "        })\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "\n",
       "    // Gaussian blur filter for making text shadow\n",
       "    var filter = svg.append('defs')\n",
       "      .append('filter')\n",
       "        .attr('id', 'text-shadow')\n",
       "        .attr('height', '130%')\n",
       "      .append('feGaussianBlur')\n",
       "        .attr('in', 'SourceAlpha')\n",
       "        .attr('stdDeviation', 2)\n",
       "        .attr('result', 'blur');\n",
       "\n",
       "    // Add invisible arcs for aligning text\n",
       "    function describeArc(radius, start, end) {\n",
       "      start_xy = polarToCartesian(radius, baseToRadians(start));\n",
       "      end_xy = polarToCartesian(radius, baseToRadians(end));\n",
       "\n",
       "      var d = [\n",
       "        'M', start_xy.x, start_xy.y,\n",
       "        'A', radius, radius, 0, 0, 1, end_xy.x, end_xy.y\n",
       "      ].join(' ');\n",
       "\n",
       "      return d;\n",
       "    };\n",
       "\n",
       "    var centerArcs = blockGroup.selectAll('.feature, .center-arcs')\n",
       "      .append('path')\n",
       "        .attr('d', function(d) {\n",
       "          return describeArc(0.97 * centerRadius, d.start, d.stop)\n",
       "        })\n",
       "        .attr('class', 'center-arc')\n",
       "        .attr('id', function(d, i) { return div_id + '-center-arc' + i; })\n",
       "        .style('stroke', 'none')\n",
       "        .style('fill', 'none');\n",
       "\n",
       "    var shadows = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-label-shadow')\n",
       "      .append('textPath')\n",
       "        .attr('class', 'shadow')\n",
       "        .attr('startOffset', '50%')\n",
       "        .attr('xlink:href', function(d, i) {\n",
       "          return '#' + div_id + '-center-arc' + i;\n",
       "        })\n",
       "        .attr('id', function(d, i) {\n",
       "          return 'shadow' + i;\n",
       "        })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '11px')\n",
       "        .style('filter', 'url(#text-shadow)')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    var blockText = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-label')\n",
       "      .append('textPath')\n",
       "        .attr('class', 'block-text')\n",
       "        .attr('fill', 'white')\n",
       "        .attr('startOffset', '50%')\n",
       "        .attr('xlink:href', function(d, i) {\n",
       "          return '#' + div_id + '-center-arc' + i;\n",
       "        })\n",
       "        .attr('id', function(d, i) { return 'label' + i; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '11px')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    // If text is too big, don't offset and set anchor to left\n",
       "    // FIXME: this should really be done with more d3-style methods\n",
       "    feature_selection = document.getElementsByClassName('feature');\n",
       "    for (var i = 0; i < feature_selection.length; i++) {\n",
       "      blocktemp = feature_selection[i];\n",
       "      text = blocktemp.getElementsByClassName('block-text')[0];\n",
       "      shadow = blocktemp.getElementsByClassName('shadow')[0];\n",
       "\n",
       "      centerArc = blocktemp.getElementsByClassName('center-arc')[0];\n",
       "      blockLength = centerArc.getTotalLength();\n",
       "\n",
       "      nameLength = text.innerHTML.length;\n",
       "      while (1.2 * text.getComputedTextLength() > blockLength) {\n",
       "         nameLength -= 1;\n",
       "         // Trim until it isn't\n",
       "         text.innerHTML = text.innerHTML.slice(0, nameLength);\n",
       "         text.setAttribute('startOffset', '10%');\n",
       "         text.style['text-anchor'] = 'start';\n",
       "\n",
       "         shadow.innerHTML = shadow.innerHTML.slice(0, nameLength);\n",
       "         shadow.setAttribute('startOffset', '10%');\n",
       "         shadow.style['text-anchor'] = 'start';\n",
       "       }\n",
       "    }\n",
       "\n",
       "    // Add line features\n",
       "    var lineGroup = featuresGroup.append('g')\n",
       "        .attr('class', 'lineFeatureGroup');\n",
       "\n",
       "    var lineFeatures = lineGroup.selectAll('.feature')\n",
       "        .data(lineData)\n",
       "      .enter().append('g')\n",
       "      .append('path')\n",
       "        .attr('class', 'feature')\n",
       "        .attr('class', 'line-feature')\n",
       "        .attr('d', function(d, i) {\n",
       "          featureAngle = baseToRadians((d.start + d.stop) / 2);\n",
       "          inner = polarToCartesian(innerRadius - featureOffset - config.tickLength, featureAngle);\n",
       "          outer = polarToCartesian(outerRadius + featureOffset + config.tickLength, featureAngle);\n",
       "\n",
       "          return line([inner, outer]);\n",
       "        })\n",
       "        .attr('stroke', 'rgb(96, 96, 96)')\n",
       "        .attr('stroke-width', config.lineFeatureWidth)\n",
       "        .attr('fill', 'none')\n",
       "        .on('mouseover', function(d) {\n",
       "          angle = baseToRadians((d.start + d.stop) / 2);\n",
       "          position = polarToCartesian((outerRadius + featureOffset + config.tickLength + 3), angle);\n",
       "          lineGroup.append('text')\n",
       "              .attr('id', 'lineFeatureHighlight')\n",
       "              .attr('x', position.x)\n",
       "              .attr('y', position.y)\n",
       "              .attr('text-anchor', 'middle')\n",
       "              .attr('font-family', 'sans-serif')\n",
       "              .attr('font-size', '14px')\n",
       "              .style('font-weight', 'bold')\n",
       "              .attr('fill', 'red')\n",
       "              .attr('dy', '-1em')\n",
       "              .attr('transform', function(d) {\n",
       "                degrees = (angle + Math.PI) / (2 * Math.PI) * 360;\n",
       "                return 'rotate(' + (degrees - 90) + ' ' + (position.x) + ',' + (position.y) + ')';\n",
       "              })\n",
       "              .text(d.name);\n",
       "        })\n",
       "        .on('mouseout', function(d) {\n",
       "          d3.select('#lineFeatureHighlight').remove();\n",
       "        });\n",
       "    } else {\n",
       "\n",
       "    ///////////////////////////////////////\n",
       "    // Placement rules for linear sequences\n",
       "    ///////////////////////////////////////\n",
       "    var backboneHeight = 25;  // TODO: replace with unified thickness param\n",
       "    var xMin = config.padding;\n",
       "    var xMax = config.width - config.padding;\n",
       "\n",
       "    var backboneDim = {\n",
       "      'left': xMin,\n",
       "      'right': xMax,\n",
       "      'top': center.y - backboneHeight / 2,\n",
       "      'bottom': center.y + backboneHeight / 2\n",
       "    };\n",
       "\n",
       "    // Offset of 1 pixel makes the display look nicer\n",
       "    var xScale = d3.scale.linear()\n",
       "        .domain([0, sequence_json.sequence.length])\n",
       "        .range([xMin + 1, xMax + 1]);\n",
       "\n",
       "    ////////////////////\n",
       "    // Draw the backbone\n",
       "    ////////////////////\n",
       "    // TODO: similarities with circular - same fill, stroke, etc. Could simplify\n",
       "    //       with a backboneWidth attribute that's in pixels\n",
       "    svg.append('rect')\n",
       "        .attr('class', 'backbone')\n",
       "        .attr('x', backboneDim.left)\n",
       "        .attr('y', backboneDim.top)\n",
       "        .attr('width', backboneDim.right - backboneDim.left)\n",
       "        .attr('height', backboneDim.bottom - backboneDim.top)\n",
       "        .style('fill', config.backboneColor)\n",
       "        .style('stroke', d3.rgb(config.backboneColor).darker(0.5))\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    /////////////////////////////////\n",
       "    // Add plasmid name, number of bp\n",
       "    /////////////////////////////////\n",
       "    // TODO: similarities with circular - identical, just different placement\n",
       "    // FIXME: make the placement of text have no magic numbers\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.name)\n",
       "        .attr('x', center.x)\n",
       "        .attr('y', backboneDim.bottom +  50)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', config.titleFontSize);\n",
       "\n",
       "    svg.append('text')\n",
       "        .text(sequence_json.sequence.length + ' bp')\n",
       "        .attr('x', center.x)\n",
       "        .attr('y', backboneDim.bottom + 80)\n",
       "        .style('text-anchor', 'middle');\n",
       "\n",
       "    /////////////\n",
       "    // Tick Marks\n",
       "    /////////////\n",
       "    // TODO: similarities with circular - almost everything\n",
       "    // Generate ticks data\n",
       "    var tickData = [];\n",
       "    for (var i=0; i < config.ticks; i++) {\n",
       "      tickData.push(i * Math.floor(sequence_json.sequence.length / config.ticks));\n",
       "    }\n",
       "\n",
       "    var tickGroup = svg.append('g')\n",
       "        .attr('id', 'tick-group');\n",
       "\n",
       "    var ticks = tickGroup.selectAll('.tick')\n",
       "        .data(tickData)\n",
       "      .enter().append('line')\n",
       "        .attr('x1', function(d) { return xScale(d); })\n",
       "        .attr('x2', function(d) { return xScale(d); })\n",
       "        .attr('y1', backboneDim.bottom)\n",
       "        .attr('y2', backboneDim.bottom + config.tickLength)\n",
       "        .attr('stroke', function(d, i) {\n",
       "          if (i % config.majorTickInterval === 0) {\n",
       "            tickColor = config.tickColorMajor;\n",
       "          } else {\n",
       "            tickColor = config.tickColorMinor;\n",
       "          }\n",
       "          return tickColor;\n",
       "        })\n",
       "        .attr('stroke-width', 2)\n",
       "        .attr('fill', 'none');\n",
       "\n",
       "    var tickLabels = tickGroup.selectAll('.tick-label')\n",
       "        .data(tickData)\n",
       "      .enter().append('text')\n",
       "        .text(function(d) {\n",
       "            if (d % config.majorTickInterval === 0) {\n",
       "              return d;\n",
       "            }\n",
       "        })\n",
       "        .attr('x', function(d) { return xScale(d); })\n",
       "        .attr('y', function(d) { return backboneDim.bottom + config.tickLength + 8; })\n",
       "        .attr('dy', '0.4em')\n",
       "        .attr('fill', config.tickColorMinor)\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-size', '12px');\n",
       "\n",
       "    ////////////////////////////////\n",
       "    // Add features and their labels\n",
       "    ////////////////////////////////\n",
       "\n",
       "    // Separate out arc features and line features\n",
       "    var blockData = [];\n",
       "    var lineData = [];\n",
       "    // FIXME: this process is broken and should be determined by fit, not n\n",
       "    // arbitrary scaling factor\n",
       "    for (var i = 0; i < sequence_json.features.length; i++) {\n",
       "      feature = sequence_json.features[i];\n",
       "      if (xScale(feature.stop - feature.start) > 10 * config.minFeatureSize) {\n",
       "        blockData.push(feature);\n",
       "      } else {\n",
       "        lineData.push(feature);\n",
       "      }\n",
       "    }\n",
       "\n",
       "    var featuresGroup = svg.append('g')\n",
       "        .attr('id', 'features');\n",
       "    var blockGroup = featuresGroup.append('g')\n",
       "        .attr('id', 'block-features');\n",
       "\n",
       "    var blockFeatures = blockGroup.selectAll('.feature')\n",
       "        .data(blockData)\n",
       "      .enter().append('g')\n",
       "        .attr('class', 'feature')\n",
       "      .append('rect')\n",
       "        .attr('class', 'feature-block')\n",
       "        .attr('x', function(d) { return xScale(d.start); })\n",
       "        .attr('y', function(d) { return backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2; })\n",
       "        .attr('width', function(d) { return xScale(d.stop) - xScale(d.start); })\n",
       "        .attr('height', function(d) { return config.featureScale * backboneHeight; })\n",
       "        .attr('id', function(d, i) { return 'block' + i; })\n",
       "        .style('fill', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            return d.color;\n",
       "          } else {\n",
       "            return 'rgb(192, 192, 256)';\n",
       "          }\n",
       "        })\n",
       "        .style('fill-opacity', config.featureOpacity)\n",
       "        .style('stroke', function(d) {\n",
       "          if (d.hasOwnProperty('color')) {\n",
       "            color = d.color;\n",
       "          } else {\n",
       "            color = 'rgb(192, 192, 256)';\n",
       "          }\n",
       "          return d3.rgb(color).darker(0.5);\n",
       "        })\n",
       "        .style('stroke-width', '1.5px');\n",
       "\n",
       "    // Gaussian blur filter for making text shadow\n",
       "    var filter = svg.append('defs')\n",
       "      .append('filter')\n",
       "        .attr('id', 'text-shadow')\n",
       "        .attr('height', '130%')\n",
       "      .append('feGaussianBlur')\n",
       "        .attr('in', 'SourceAlpha')\n",
       "        .attr('stdDeviation', 2)\n",
       "        .attr('result', 'blur');\n",
       "\n",
       "    var shadows = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'shadow')\n",
       "        .attr('id', function(d, i) {\n",
       "          return 'shadow' + i;\n",
       "        })\n",
       "        .attr('x', function(d) { return xScale((d.start + d.stop) / 2); })\n",
       "        .attr('y', function(d) { return (backboneDim.top + backboneDim.bottom) / 2 + 3; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '10px')\n",
       "        .style('filter', 'url(#text-shadow)')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    var blockText = blockGroup.selectAll('.feature')\n",
       "      .append('text')\n",
       "        .attr('class', 'block-text')\n",
       "        .attr('fill', 'white')\n",
       "        .attr('id', function(d, i) { return 'label' + i; })\n",
       "        .attr('x', function(d) { return xScale((d.start + d.stop) / 2); })\n",
       "        .attr('y', function(d) { return (backboneDim.top + backboneDim.bottom) / 2; })\n",
       "        .style('text-anchor', 'middle')\n",
       "        .style('font-family', 'sans-serif')\n",
       "        .style('font-weight', 'bold')\n",
       "        .style('font-size', '10px')\n",
       "        .text(function(d) { return d.name; });\n",
       "\n",
       "    // If text is too big, don't offset and set anchor to left\n",
       "    // FIXME: this should really be done with more d3-style methods\n",
       "    feature_selection = document.getElementsByClassName('feature');\n",
       "    for (var i = 0; i < feature_selection.length; i++) {\n",
       "      blocktemp = feature_selection[i];\n",
       "      text = blocktemp.getElementsByClassName('block-text')[0];\n",
       "      shadow = blocktemp.getElementsByClassName('shadow')[0];\n",
       "\n",
       "      block = blocktemp.getElementsByClassName('feature-block');\n",
       "      blockLength = block[0].width.baseVal.value\n",
       "\n",
       "      nameLength = text.innerHTML.length;\n",
       "      while (1.2 * text.getComputedTextLength() > blockLength) {\n",
       "         nameLength -= 1;\n",
       "         // Trim until it isn't\n",
       "         text.innerHTML = text.innerHTML.slice(0, nameLength);\n",
       "\n",
       "         shadow.innerHTML = shadow.innerHTML.slice(0, nameLength);\n",
       "       }\n",
       "    }\n",
       "\n",
       "    // Add line features\n",
       "    var lineGroup = featuresGroup.append('g')\n",
       "        .attr('class', 'lineFeatureGroup');\n",
       "\n",
       "    var lineFeatures = lineGroup.selectAll('.feature')\n",
       "        .data(lineData)\n",
       "      .enter().append('g')\n",
       "      .append('line')\n",
       "        .attr('class', 'feature')\n",
       "        .attr('class', 'line-feature')\n",
       "        .attr('x1', function(d) { return xScale(d.start); })\n",
       "        .attr('x2', function(d) { return xScale(d.start); })\n",
       "        .attr('y1', function(d) { return backboneDim.bottom + (config.featureScale * backboneHeight - backboneHeight) / 2 + config.tickLength; })\n",
       "        .attr('y2', function(d) { return backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2 - config.tickLength; })\n",
       "        .attr('stroke', 'rgb(96, 96, 96)')\n",
       "        .attr('stroke-width', config.lineFeatureWidth)\n",
       "        .attr('fill', 'none')\n",
       "        .on('mouseover', function(d) {\n",
       "          lineGroup.append('text')\n",
       "              .attr('id', 'lineFeatureHighlight')\n",
       "              .attr('x', xScale(d.start))\n",
       "              .attr('y', backboneDim.top - (config.featureScale * backboneHeight - backboneHeight) / 2 - 3)\n",
       "              .attr('text-anchor', 'middle')\n",
       "              .attr('font-family', 'sans-serif')\n",
       "              .attr('font-size', '14px')\n",
       "              .style('font-weight', 'bold')\n",
       "              .attr('fill', 'red')\n",
       "              .attr('dy', '-1em')\n",
       "              .text(d.name);\n",
       "        })\n",
       "        .on('mouseout', function(d) {\n",
       "          d3.select('#lineFeatureHighlight').remove();\n",
       "        });\n",
       "\n",
       "  }\n",
       "}\n",
       "\n",
       "            d3sequence(window.data, 'sequence_e1200884-c47f-11e5-9eaa-02591d3bb8ad')\n",
       "        });\n",
       "        </script>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mockup = (gene + backbone).circularize()\n",
    "mockup\n",
    "mockup.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[YJL014W 'CDS' feature (0 to 1605) on strand 0,\n",
       " PP2 'primer_bind' feature (6051 to 6075) on strand 0,\n",
       " Actual GPD Promoter 'promoter' feature (6075 to 6730) on strand 0,\n",
       " TP 'misc_feature' feature (1605 to 1627) on strand 0,\n",
       " CYC1_terminator 'terminator' feature (1627 to 1867) on strand 0,\n",
       " pYESTrp_rev_primer 'misc_feature' feature (1636 to 1655) on strand 0,\n",
       " CYC1_primer 'misc_feature' feature (1636 to 1655) on strand 0,\n",
       " TS 'primer_bind' feature (1867 to 1889) on strand 0,\n",
       " URA3 3' UTR '3'UTR' feature (1889 to 2389) on strand 0,\n",
       " PmeI(1) 'misc_feature' feature (2389 to 2397) on strand 1,\n",
       " M13 Reverse primer 'promoter' feature (2401 to 2418) on strand 0,\n",
       " M13_pUC_rev_primer 'misc_feature' feature (2414 to 2437) on strand 0,\n",
       " mutant? 'misc_feature' feature (2456 to 2457) on strand 1,\n",
       " pBR322_origin 'rep_origin' feature (2788 to 3408) on strand 0,\n",
       " mut? 'misc_feature' feature (2816 to 2817) on strand 1,\n",
       " Ampicillin 'CDS' feature (3562 to 4423) on strand 1,\n",
       " AmpR_promoter 'promoter' feature (4464 to 4493) on strand 1,\n",
       " pGEX_3_primer 'misc_feature' feature (4651 to 4674) on strand 0,\n",
       " PmeI 'misc_feature' feature (4809 to 4817) on strand 0,\n",
       " URA3 promoter (kind of) 'promoter' feature (4817 to 5065) on strand 0,\n",
       " URA3 'gene' feature (5065 to 5770) on strand 0,\n",
       " mutated to A in ura3-1 strains 'misc_difference' feature (5765 to 5766) on strand 0,\n",
       " ADH1 Terminator 'terminator' feature (5869 to 6051) on strand 0]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mockup.features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3: Design the Gibson primers necessary to build the plasmid\n",
    "\n",
    "The `design.gibson_primers` function designs the two primers that would create a 'Gibson junction' (homology) on two separate fragments - i.e., it designs a reverse primer for fragment 1 and a forward primer for fragment 2.\n",
    "\n",
    "The `overlap` keyword argument tells the design algorithm to put all of the overhang on just one primer (the one for the gene, in our case)\n",
    "\n",
    "Note that the `gibson_primers` function creates `coral.Primer` objects, which are containers for the annealing sequence (lowercase), overhang sequence (uppercase), and a specified Tm attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Primer: atcggattctagaactagtggatctacaaaATGCAAGCTCCAGTGGTATTTATG Tm: 65.15,\n",
       " Primer: actcgaggtcgacggtatcaTTATTCTTGTTTTCTTACACCACTAACAATATC Tm: 64.19]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We choose to put all of the overlap on the gene primers, since backbone\n",
    "# fragments already exist (we'll use those so as to minimize cost + time)\n",
    "gene_rev, backbone_fwd = cr.design.gibson_primers(gene, backbone, overlap='right')\n",
    "backbone_rev, gene_fwd = cr.design.gibson_primers(backbone, gene, overlap='left')\n",
    "\n",
    "# Note that we don't need to use or order the backbone primers,\n",
    "# because we already have primers that do that (and their fragments)\n",
    "gene_primers = [gene_fwd, gene_rev]\n",
    "gene_primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Validation of design ('reverse' process)\n",
    "\n",
    "### 1: Simulate the PCR on the templates to be used, generates fragments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Our gene is on chromosome 10 of the yeast genome\n",
    "chr10 = cr.seqio.read_dna('./chr10.gb')\n",
    "gene_amplicon = cr.reaction.pcr(chr10, gene_primers[0], gene_primers[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### 2: Simulate the Gibson reaction on the fragments, generates plasmid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gibsoned = cr.reaction.gibson([gene_amplicon, backbone])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3: Compare simulated constructed plasmid to mockup."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Why is the first comparison False, and the second True?\n",
    "print gibsoned == mockup\n",
    "print gibsoned.is_rotation(mockup)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Aquarium submission process:\n",
    "\n",
    "### 1. Log into rehearsal server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import aquariumapi\n",
    "\n",
    "# The :82 is important! This is where the rehearsal server works.\n",
    "# NEVER test API-using code on the production server\n",
    "api = aquariumapi.AquariumAPI('http://54.68.9.194:82/api', 'bolten', 'yourkey')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Locate template(s) for PCR (thing that already exists on Aquarium)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{u'AssemblyOligo': AssemblyOligo SampleModel,\n",
       " u'AssemblyPool': AssemblyPool SampleModel,\n",
       " u'E coli strain': E coli strain SampleModel,\n",
       " u'Enzyme': Enzyme SampleModel,\n",
       " u'Enzyme Buffer': Enzyme Buffer SampleModel,\n",
       " u'Fragment': Fragment SampleModel,\n",
       " u'Inducer': Inducer SampleModel,\n",
       " u'Ladder': Ladder SampleModel,\n",
       " u'Plasmid': Plasmid SampleModel,\n",
       " u'Primer': Primer SampleModel,\n",
       " u'Primer Pool': Primer Pool SampleModel,\n",
       " u'Tetra': Tetra SampleModel,\n",
       " u'Transformed E coli Strain': Transformed E coli Strain SampleModel,\n",
       " u'Yeast Strain': Yeast Strain SampleModel}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# The template is a yeast strain. \n",
    "# We can use almost any strain in our lab, so we chose\n",
    "# the yeast strain with sample ID 12103\n",
    "\n",
    "# We want to work with database models in Aquarium.\n",
    "# aquariumapi supplies a module for working more easily\n",
    "# with those models\n",
    "sample_defs = aquariumapi.models.get_sample_definitions(api)\n",
    "display(sample_defs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{u'result': u'ok',\n",
       " u'rows': [{u'created_at': u'2013-10-22T06:33:30-07:00',\n",
       "   u'data': None,\n",
       "   u'description': u'Cultures associated with this glycerol stock of W303a are contaminated, grow poorly, and have low transformation efficiency.',\n",
       "   u'fields': {u'Integrant': u'-none-',\n",
       "    u'Integrated Marker(s)': u'',\n",
       "    u'Mating Type': u'a',\n",
       "    u'Parent': u'-none-',\n",
       "    u'Plasmid': u'-none-',\n",
       "    u'Plasmid Marker(s)': u'',\n",
       "    u'QC Primer1': u'TCYC1_F',\n",
       "    u'QC Primer2': u'ColonyPCR_URA3_R(pMOD)'},\n",
       "   u'id': 29,\n",
       "   u'name': u'W303a faulty',\n",
       "   u'project': u'Yeast General',\n",
       "   u'sample_type_id': 5,\n",
       "   u'updated_at': u'2015-09-01T04:46:11-07:00',\n",
       "   u'user_id': 16}],\n",
       " u'warnings': []}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# We want to work with a yeast strain\n",
    "yeastmodel = sample_defs['Yeast Strain']\n",
    "query = yeastmodel.find({'id': 29})\n",
    "display(query)\n",
    "yeastname = query['rows'][0]['name']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Submit primers to Aquarium (save the IDs!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Overhang Sequence', 'string'],\n",
       " ['Anneal Sequence', 'string'],\n",
       " ['T Anneal', 'number']]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# What info is needed to submit primers?\n",
    "primermodel = sample_defs['Primer']\n",
    "display(primermodel.fields)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Name the primers in advance (and supply metadata)\n",
    "gene_primers[0].name = gene.name + '_Fwd'\n",
    "gene_primers[0].gene = gene.name\n",
    "gene_primers[1].name = gene.name + '_Rev'\n",
    "gene_primers[1].gene = gene.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<type 'float'> 65.1547148567\n",
      "<type 'float'> 64.1877903446\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{u'result': u'ok',\n",
       "  u'rows': [{u'created_at': u'2016-01-26T14:54:39-08:00',\n",
       "    u'data': None,\n",
       "    u'description': u'Primer to amplify YJL014W with pMOD overhangs.',\n",
       "    u'field1': u'ATCGGATTCTAGAACTAGTGGATCTACAAA',\n",
       "    u'field2': u'ATGCAAGCTCCAGTGGTATTTATG',\n",
       "    u'field3': None,\n",
       "    u'field4': None,\n",
       "    u'field5': None,\n",
       "    u'field6': None,\n",
       "    u'field7': None,\n",
       "    u'field8': None,\n",
       "    u'id': 11799,\n",
       "    u'name': u'YJL014W_Fwd',\n",
       "    u'project': u'LABW16',\n",
       "    u'properties': {u'': None,\n",
       "     u'Anneal Sequence': u'ATGCAAGCTCCAGTGGTATTTATG',\n",
       "     u'Overhang Sequence': u'ATCGGATTCTAGAACTAGTGGATCTACAAA',\n",
       "     u'T Anneal': 0},\n",
       "    u'sample_type_id': 1,\n",
       "    u'updated_at': u'2016-01-26T14:54:39-08:00',\n",
       "    u'user_id': 4}],\n",
       "  u'warnings': []},\n",
       " {u'errors': [u'Could not create Sample: Name Samples: must have unique names.'],\n",
       "  u'result': u'error'}]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Submit the primers\n",
    "# IMPORTANT: only run this once if you can\n",
    "submitted_primers = []\n",
    "for primer in gene_primers:\n",
    "    desc = 'Primer to amplify {} with pMOD overhangs.'.format(primer.gene)\n",
    "    print type(primer.tm), primer.tm\n",
    "    fields = {'Anneal Sequence': str(primer.anneal),\n",
    "              'Overhang Sequence': str(primer.overhang)}\n",
    "    submitted = primermodel.create(primer.name, \n",
    "                                   desc,\n",
    "                                   fields,\n",
    "                                   'LABW16')\n",
    "    submitted_primers.append(submitted)\n",
    "submitted_primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Submit fragments to Aquarium using template + fragment IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Sequence', 'url'],\n",
       " ['Length', 'number'],\n",
       " ['Template', 'Plasmid|E coli strain|Fragment|Yeast Strain'],\n",
       " ['Forward Primer', 'Primer'],\n",
       " ['Reverse Primer', 'Primer'],\n",
       " ['Restriction Enzyme(s)', 'string'],\n",
       " ['Yeast Marker', 'string']]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragment_model = sample_defs['Fragment']\n",
    "fragment_model.fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{u'result': u'ok',\n",
       " u'rows': [{u'created_at': u'2016-01-26T14:56:47-08:00',\n",
       "   u'data': None,\n",
       "   u'description': u'YJL014W with pMOD overhangs',\n",
       "   u'field1': None,\n",
       "   u'field2': 1655,\n",
       "   u'field3': u'W303a faulty',\n",
       "   u'field4': u'YJL014W_Fwd',\n",
       "   u'field5': u'YJL014W_Rev',\n",
       "   u'field6': None,\n",
       "   u'field7': None,\n",
       "   u'field8': None,\n",
       "   u'id': 11800,\n",
       "   u'name': u'YJL014W',\n",
       "   u'project': u'LABW16',\n",
       "   u'properties': {u'': None,\n",
       "    u'Forward Primer': {u'created_at': u'2016-01-26T14:54:39-08:00',\n",
       "     u'data': None,\n",
       "     u'description': u'Primer to amplify YJL014W with pMOD overhangs.',\n",
       "     u'field1': u'ATCGGATTCTAGAACTAGTGGATCTACAAA',\n",
       "     u'field2': u'ATGCAAGCTCCAGTGGTATTTATG',\n",
       "     u'field3': None,\n",
       "     u'field4': None,\n",
       "     u'field5': None,\n",
       "     u'field6': None,\n",
       "     u'field7': None,\n",
       "     u'field8': None,\n",
       "     u'id': 11799,\n",
       "     u'name': u'YJL014W_Fwd',\n",
       "     u'project': u'LABW16',\n",
       "     u'properties': {u'': None,\n",
       "      u'Anneal Sequence': u'ATGCAAGCTCCAGTGGTATTTATG',\n",
       "      u'Overhang Sequence': u'ATCGGATTCTAGAACTAGTGGATCTACAAA',\n",
       "      u'T Anneal': 0},\n",
       "     u'sample_type_id': 1,\n",
       "     u'updated_at': u'2016-01-26T14:54:39-08:00',\n",
       "     u'user_id': 4},\n",
       "    u'Length': 1655,\n",
       "    u'Restriction Enzyme(s)': None,\n",
       "    u'Reverse Primer': {u'created_at': u'2016-01-26T14:54:02-08:00',\n",
       "     u'data': None,\n",
       "     u'description': u'Primer to amplify YJL014W with pMOD overhangs.',\n",
       "     u'field1': u'ACTCGAGGTCGACGGTATCA',\n",
       "     u'field2': u'TTATTCTTGTTTTCTTACACCACTAACAATATC',\n",
       "     u'field3': u'64.1877903446121',\n",
       "     u'field4': None,\n",
       "     u'field5': None,\n",
       "     u'field6': None,\n",
       "     u'field7': None,\n",
       "     u'field8': None,\n",
       "     u'id': 11798,\n",
       "     u'name': u'YJL014W_Rev',\n",
       "     u'project': u'LABW16',\n",
       "     u'properties': {u'': None,\n",
       "      u'Anneal Sequence': u'TTATTCTTGTTTTCTTACACCACTAACAATATC',\n",
       "      u'Overhang Sequence': u'ACTCGAGGTCGACGGTATCA',\n",
       "      u'T Anneal': 64.1877903446121},\n",
       "     u'sample_type_id': 1,\n",
       "     u'updated_at': u'2016-01-26T14:54:02-08:00',\n",
       "     u'user_id': 4},\n",
       "    u'Sequence': None,\n",
       "    u'Template': {u'created_at': u'2013-10-22T06:33:30-07:00',\n",
       "     u'data': None,\n",
       "     u'description': u'Cultures associated with this glycerol stock of W303a are contaminated, grow poorly, and have low transformation efficiency.',\n",
       "     u'field1': u'-none-',\n",
       "     u'field2': u'-none-',\n",
       "     u'field3': u'-none-',\n",
       "     u'field4': u'',\n",
       "     u'field5': u'',\n",
       "     u'field6': u'a',\n",
       "     u'field7': u'TCYC1_F',\n",
       "     u'field8': u'ColonyPCR_URA3_R(pMOD)',\n",
       "     u'id': 29,\n",
       "     u'name': u'W303a faulty',\n",
       "     u'project': u'Yeast General',\n",
       "     u'properties': {u'Integrant': None,\n",
       "      u'Integrated Marker(s)': u'',\n",
       "      u'Mating Type': u'a',\n",
       "      u'Parent': None,\n",
       "      u'Plasmid': None,\n",
       "      u'Plasmid Marker(s)': u'',\n",
       "      u'QC Primer1': {u'created_at': u'2014-08-27T09:31:18-07:00',\n",
       "       u'data': None,\n",
       "       u'description': u\"Cloning/Sequencing primer that binds at the beginning of the CYC1 terminator sequence in pAG/pGP vectors. Note: it doesn't perfectly match the SGD genome. Migrated from primer 606, designed by Nick Bolten.\",\n",
       "       u'field1': u'',\n",
       "       u'field2': u'ATTAGTTATGTCACGCTTACATTCACGCCC',\n",
       "       u'field3': u'71.83',\n",
       "       u'field4': None,\n",
       "       u'field5': None,\n",
       "       u'field6': None,\n",
       "       u'field7': None,\n",
       "       u'field8': None,\n",
       "       u'id': 2053,\n",
       "       u'name': u'TCYC1_F',\n",
       "       u'project': u'Bandpass Filter',\n",
       "       u'properties': {u'': None,\n",
       "        u'Anneal Sequence': u'ATTAGTTATGTCACGCTTACATTCACGCCC',\n",
       "        u'Overhang Sequence': u'',\n",
       "        u'T Anneal': 71.83},\n",
       "       u'sample_type_id': 1,\n",
       "       u'updated_at': u'2014-08-27T09:31:18-07:00',\n",
       "       u'user_id': 20},\n",
       "      u'QC Primer2': {u'created_at': u'2014-09-09T02:05:45-07:00',\n",
       "       u'data': None,\n",
       "       u'description': None,\n",
       "       u'field1': u'',\n",
       "       u'field2': u'GGTTTCTCTTTATATTTACATGCTAAAAATGGGCTACACC',\n",
       "       u'field3': u'70',\n",
       "       u'field4': None,\n",
       "       u'field5': None,\n",
       "       u'field6': None,\n",
       "       u'field7': None,\n",
       "       u'field8': None,\n",
       "       u'id': 2149,\n",
       "       u'name': u'ColonyPCR_URA3_R(pMOD)',\n",
       "       u'project': u'DY Yeast General',\n",
       "       u'properties': {u'': None,\n",
       "        u'Anneal Sequence': u'GGTTTCTCTTTATATTTACATGCTAAAAATGGGCTACACC',\n",
       "        u'Overhang Sequence': u'',\n",
       "        u'T Anneal': 70},\n",
       "       u'sample_type_id': 1,\n",
       "       u'updated_at': u'2014-09-09T02:05:45-07:00',\n",
       "       u'user_id': 13}},\n",
       "     u'sample_type_id': 5,\n",
       "     u'updated_at': u'2015-09-01T04:46:11-07:00',\n",
       "     u'user_id': 16},\n",
       "    u'Yeast Marker': None},\n",
       "   u'sample_type_id': 4,\n",
       "   u'updated_at': u'2016-01-26T14:56:47-08:00',\n",
       "   u'user_id': 4}],\n",
       " u'warnings': []}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fields = {'Length': len(gene_amplicon),\n",
    "          'Template': yeastname,\n",
    "          'Forward Primer': gene_primers[0].name,\n",
    "          'Reverse Primer': gene_primers[1].name}\n",
    "\n",
    "fragment_model.create(gene.name,\n",
    "                     '{} with pMOD overhangs'.format(gene.name),\n",
    "                     fields,\n",
    "                     'LABW16')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Submit plasmid to Aquarium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Sequence', 'url'],\n",
       " ['Sequence Verification', 'url'],\n",
       " ['Bacterial Marker', 'string'],\n",
       " ['Yeast Marker', 'string'],\n",
       " ['Length', 'number'],\n",
       " ['Sequencing_primer_ids', 'string']]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plasmidmodel = sample_defs['Plasmid']\n",
    "plasmidmodel.fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{u'result': u'ok',\n",
       " u'rows': [{u'created_at': u'2016-01-26T14:58:16-08:00',\n",
       "   u'data': None,\n",
       "   u'description': u'Overexpression plasmid for YJL014W',\n",
       "   u'field1': None,\n",
       "   u'field2': None,\n",
       "   u'field3': u'AmpR',\n",
       "   u'field4': u'URA3',\n",
       "   u'field5': 6758,\n",
       "   u'field6': None,\n",
       "   u'field7': None,\n",
       "   u'field8': None,\n",
       "   u'id': 11801,\n",
       "   u'name': u'pMOD6G-YJL014W',\n",
       "   u'project': u'LABW16',\n",
       "   u'properties': {u'': None,\n",
       "    u'Bacterial Marker': u'AmpR',\n",
       "    u'Length': 6758,\n",
       "    u'Sequence': None,\n",
       "    u'Sequence Verification': None,\n",
       "    u'Sequencing_primer_ids': None,\n",
       "    u'Yeast Marker': u'URA3'},\n",
       "   u'sample_type_id': 2,\n",
       "   u'updated_at': u'2016-01-26T14:58:16-08:00',\n",
       "   u'user_id': 4}],\n",
       " u'warnings': []}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fields = {'Bacterial Marker': 'AmpR',\n",
    "          'Yeast Marker': 'URA3',\n",
    "          'Length': len(mockup)}\n",
    "\n",
    "plasmidmodel.create('pMOD6G-{}'.format(gene.name),\n",
    "                    'Overexpression plasmid for {}'.format(gene.name),\n",
    "                    fields,\n",
    "                    'LABW16')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
